---
title: "Mash Sketch for Short Reads"
---

```{r mash-setup, include=FALSE}
write.table(c("",
              "####################################",
              "####            Mash            ####",
              "####################################"),
            file=fn_log, quote=F, row.names=F, col.names=F, append=T)

# refseq and short reads directories
dir_refseq <- paste0(getwd(), "/refseq/")
dir_shortreads <- paste0(getwd(), "/short_reads/")

# lists of refseq and short reads
ls_refseq <- list.dirs(dir_refseq, recursive=F, full.names=F)
ls_shortreads <- list.dirs(dir_shortreads, recursive=F, full.names=F)

# output files
file_msh <- paste0(dir_refseq, params$prefix, ".msh")
file_mashsum <- paste0(params$outdir, "/", params$prefix, ".mashsum")
```

## Create Mash sketch for all reference sequences
```{r mash-sketch}
write.table(c("",
              "--------- Mash Sketch on References --------"),
              file=fn_log, quote=F, row.names=F, col.names=F, append=T)

# list of reference FASTA sequences
ls_refseq_fasta <- NULL
ls_refseq_fasta <- sapply(ls_refseq, function(x) {
    dir_fasta <- paste0(dir_refseq, x, "/ncbi_dataset/data/", x, "/")
    file_fasta <- list.files(dir_fasta, pattern="*.fna", recursive=F)

    if (length(file_fasta) > 0) {
        c(ls_refseq_fasta, file_fasta)
    }
})

# check the length of the reference sequences
if (length(ls_refseq_fasta) == 0 || is.null(ls_refseq_fasta)) {
    log4r::error(fn_logger, "Error: list of reference sequence has length zero. Exited.")
    knitr::knit_exit()
} else {
    # convert list to string
    str_refseq_fasta <- paste(ls_refseq_fasta, collapse=" ")

    # check if output file exists
    prefix <- paste0(dir_refseq, params$prefix)
    if (file.exists(file_msh) && !params$redo) {
        log4r::warn(fn_logger, "File found: Mash sketch for reference sequences. Skipped.")
    } else {
        # run Mash sketch on all reference sequences
        f_mash_sketch(params$exe_mash, str_refseq_fasta, params$thread, prefix)
        log4r::info(fn_logger, "File created: Mash sketch for reference sequences.")
    }
}
```

## Run Mash screen for all short reads
```{r mash-screen}
write.table(c("",
              "--------- Mash Screen on Short Reads -------"),
              file=fn_log, quote=F, row.names=F, col.names=F, append=T)

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

foreach (i = ls_shortreads) %dopar% {
    dir_fastq <- paste0(dir_shortreads, i, "/fastq/")

    # extract the forward fastq
    fwd_read <- list.files(dir_fastq, full.names=T)[1]
    if (length(fwd_read) == 0) {
        log4r::warn(fn_logger, paste0("File not found: FASTQ alignment for ", i, ". Skipped."))
        return(NULL)
    }

    file_output <- paste0(dir_shortreads, i, "/", i, ".mashtab")
    if (file.exists(file_output) && !params$redo) {
        log4r::warn(fn_logger, paste0("File found: .mashtab for ", i, ". Skipped."))
        return(NULL)
    }

    f_mash_screen(params$exe_mash, file_msh, fwd_read, file_output)
    log4r::info(fn_logger, paste0("File created: .mashtab for ", i, "."))
}

stopCluster(nwcl)
```

## Summarize .mashtab for all short reads
```{r mash-screen-summary}
df_mashsum <- data.table::data.table(reads_accession=character(), reads_dir=character(), reads_species=character(),
                                    refseq_accession=character(), refseq_dir=character(), refseq_species=character(),
                                    score=numeric())

for (i in ls_shortreads) {
    # input files
    reads_accession <- i
    reads_dir <- list.files(paste0(dir_shortreads, i, "/fastq/"), full.names=T)[1]
    reads_species <- f_extract_species_from_reads(params$exe_esearch, params$efetch, i)

    file_mashtab <- paste0(dir_shortreads, i, "/", i, ".mashtab")
    if (!file.exists(file_output)) {
        log4r::warn(fn_logger, paste0("File not found: .mashtab for ", i, ". Skipped."))
        return(NULL)
    }

    # extract the best Mash sketch
    cmd_sort_mashtab <- paste("sort -gr", file_mashtab, "| head -1")
    best_sketch <- system(cmd_sort_mashtab, intern=T)
    ls_best_sketch <- unlist(strsplit(best_sketch, split="\t"))

    # extract the accession number
    refseq_dir <- ls_best_sketch[5]
    ls_refseq_dir <- unlist(strsplit(refseq_dir, split="/"))
    ls_refseq_accession <- unlist(strsplit(ls_refseq_dir[length(ls_refseq_dir)], split="_"))
    refseq_accession <- paste(ls_refseq_accession[1:2], collapse="_")

    # extract the species name and score
    refseq_species <- f_extract_species_from_assembly(params$exe_esearch, params$efetch, params$xtract, refseq_accession)
    score <- as.numeric(ls_best_sketch[1])
    
    # save the information into data.table
    df_mashsum <- rbind(df_mashsum, list(reads_accession=reads_accession, reads_dir=reads_dir, reads_species=reads_species,
                                    refseq_accession=refseq_accession, refseq_dir=refseq_dir, refseq_species=refseq_species,
                                    score=score))
}

data.table::fwrite(df_mashsum, file=file_mashsum, quote=F, sep="\t")
log4r::info(fn_logger, "File created/updated: summary file for Mash Sketches.")
```