---
title: "Run BUSCO Analysis on Reference Sequences and Mapped Reads"
---

```{r check-busco-setup, include=FALSE}
f_write_log(fn_log=fn_log,
            msg=c("", "####################################",
                      "####            BUSCO           ####",
                      "####################################"))

# input files
fn_readmap_metadata <- paste0(currentdir, "/readmap/metadata.tsv")
df_refseq_dir <- data.table::fread(fn_readmap_metadata)

df_refs <- data.table::fread(params$file_refseq)
ls_refseq <- df_refs$id

df_reads <- data.table::fread(params$file_shortreads)
ls_shortreads <- df_reads$id[df_reads$is_check]

# output directories
dir_busco <- ""

# check the BUSCO extraction method
if (params$type == "coordinate" || params$type == "" || is.null(params$type)) {
    dir_busco <- paste0(currentdir, "/busco_check/busco_coordinate/")
    f_write_log(fn_log=fn_log, msg="Running coordinate-based method.")
} else if (params$type == "pipeline") {
    dir_busco <- paste0(currentdir, "/busco_check/busco_pipeline/")
    f_write_log(fn_log=fn_log, msg="Running pipeline-based method.")
} else {
    log4r::error(fn_logger, "Invalid type of BUSCO extraction method. Exited.")
    knitr::knit_exit()
}

dir_busco_lineage <- paste0(currentdir, "/busco_lineage/")
dir_busco_ref <- paste0(currentdir, "/busco_check/busco_refseq/")

dir_busco_read <- paste0(dir_busco, "short_reads/")
dir_busco_tree <- paste0(dir_busco, "trees/")

lapply(list(dir_busco_lineage,dir_busco_ref,dir_busco_read,dir_busco_tree), function(x){if(!dir.exists(x)) dir.create(x, recursive=T)})

# output files
fn_refs_busco <- paste0(dir_busco_ref, "metadata.tsv")
fn_mapped_reads_busco <- paste0(dir_busco_read, "metadata.tsv")
```

```{r check-busco-params, include=FALSE}
min_ref_busco_identity <- 100

thread_busco <- ifelse(params$thread < 50, params$thread, 50)
```

```{r busco-check-species-map, include=FALSE}
# create a map for ID and species name
ls_species_name <- c()

# iterate over reference sequences
for (ref in 1:nrow(df_refs)) {
    ls_species_name[df_refs$id[ref]] <- df_refs$species[ref]
}

# iterate over short reads
for (read in 1:nrow(df_reads)) {
    ls_species_name[df_reads$id[read]] <- df_reads$species[read]
}
```

## Download BUSCO lineage dataset
```{r busco-lineage-download}
# download specific BUSCO lineage
dir_lineage <- paste0(dir_busco_lineage, params$busco_lineage)
if (!dir.exists(dir_lineage)) {
    busco_cmd <- paste(params$exe_busco, "--download_path", dir_busco_lineage, "--download", params$busco_lineage)
    system(busco_cmd)
}
```

## Run BUSCO on reference sequences
```{r, include=FALSE}
f_write_log(fn_log=fn_log, msg=c("", "--------- Reference Sequence -------"))

# set up variables
is_busco_ref <- FALSE

ls_busco_ref_done <- c()
ls_busco_ref_run <- c()

# iterate over reference sequences
if (params$redo) {
    ls_busco_ref_run <- ls_refseq
} else {
    for (ref in ls_refseq) {
        # check if BUSCO run exists
        file_busco_log <- paste0(dir_busco_ref, ref, "/logs/busco.log")
        if (file.exists(file_busco_log)) {
            finish_msg <- system(paste("grep 'BUSCO analysis done'", file_busco_log), intern=T)
            if (length(finish_msg) != 0) {
                ls_busco_ref_done <- c(ls_busco_ref_done, paste("-", ref))
                next
            }
        }

        # add reference sequence to to-do-list
        ls_busco_ref_run <- c(ls_busco_ref_run, ref)
    }
}

# output log file for available reference BUSCO runs
if (length(ls_busco_ref_done) > 0) {
    ls_busco_ref_done <- c(paste0("Available reference BUSCO runs (", length(ls_busco_ref_done), "/", length(ls_refseq), ")"), ls_busco_ref_done, "")
    f_write_log(fn_log=fn_log, msg=ls_busco_ref_done)
}

# update variable for the next analysis
if (length(ls_busco_ref_run) > 0) {
    is_busco_ref <- TRUE
}
```

```{r}
# print the number of reference BUSCO in HTML document
cat(paste0("Available reference BUSCO (", length(ls_busco_ref_done), "/", length(ls_refseq), ")"))
cat(paste("Reference BUSCO to run:", length(ls_busco_ref_run)))
```

```{r check-busco-ref, include=is_busco_ref, eval=is_busco_ref}
f_write_log(fn_log=fn_log, msg=paste("Running BUSCO using", thread_busco, "threads"))

# create doSNOW cluster
nwcl <- makeCluster(floor(params$thread/thread_busco))
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
foreach (ref = ls_refseq) %dopar% {
    # extract reference file location
    file_ref <- df_refseq_dir$dir[df_refseq_dir$id==ref]

    # run BUSCO
    f_run_busco(file_ref, dir_lineage, ref, dir_busco_ref, params$busco_mode, thread_busco, params$exe_busco)
    log4r::info(fn_logger, paste0("File created: BUSCO run for ", ref, "."))
}

stopCluster(nwcl)
```

## Extract common BUSCOs between reference sequences
```{r check-busco-common}
# initiate variable
shared_busco <- c()

# iterate over reference sequences
for (ref in ls_refseq) {
  # directory for BUSCO alignments
  dir_buscos <- paste0(dir_busco_ref, ref, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/")
  
  # remove file extension .faa
  busco_ids <- list.files(dir_buscos, pattern = "*.faa$", full.names = F, recursive = F)
  busco_ids <- sapply(busco_ids, function(x) { gsub(".faa", "", x) })
  
  # extract shared BUSCOs between reference sequences
  if (length(shared_busco) == 0) {
    shared_busco <- busco_ids
  } else {
    shared_busco <- intersect(shared_busco, busco_ids)
  }
}

# print the number of BUSCO in HTML document
cat(paste("Number of shared BUSCO between references:", length(shared_busco)))

# check if shared BUSCOs found
if (length(shared_busco) == 0) {
    log4r::error(fn_logger, paste0("Error: no BUSCO shared between reference sequences. Exited."))
    knitr::knit_exit()
} else {
    f_write_log(fn_log=fn_log, msg=c("", paste0("Number of complete single-copy BUSCOs: ", length(shared_busco), ".")))
}
```

## Extract BUSCOs for reference sequences
```{r check-busco-extract-ref-metadata, include=FALSE}
is_ref_extract <- TRUE

# open the metadata file
if (file.exists(fn_refs_busco) && !params$redo) {
    df_chr_busco <- data.table::fread(fn_refs_busco)

    # exclude problematic BUSCOs from further analyses
    shared_busco <- unique(df_chr_busco$busco)
    ls_problematic_busco <- unique(df_chr_busco$busco[df_chr_busco$is_error])
    shared_busco <- shared_busco[!shared_busco %in% ls_problematic_busco]

    # print the number of BUSCO in HTML document
    cat(paste("Number of shared BUSCO after filtering non-identical genes:", length(shared_busco)))
    f_write_log(fn_log=fn_log, msg=c("", paste0("Number of BUSCO after filtering problematic genes: ", length(shared_busco), ".")))

    is_ref_extract <- FALSE
}

# stop the knit if shared busco is zero
if (length(shared_busco) == 0) {
    knitr::knit_exit()
}
```

```{r check-busco-extract-ref, include=is_ref_extract, eval=is_ref_extract}
# output data.table
df_chr_busco <- data.table::data.table(ref=character(), busco=character(), is_error=logical(), identity=numeric())

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
for (ref in ls_refseq) {
    # directory for BUSCO alignments
    dir_busco_out <- paste0(dir_busco_ref, ref, "/fasta/")
    if (!dir.exists(dir_busco_out)) {
        dir.create(dir_busco_out, recursive=T)
    }
  
    # extract reference file location
    file_fas <- df_refseq_dir$dir[df_refseq_dir$id==ref]

    # iterate over shared BUSCOs
    df_output_temp <- foreach (busco = shared_busco, .combine='rbind') %dopar% {
        require(data.table)
        require(Biostrings)
        require(stringr)

        # BUSCO files
        dir_busco_gff <- paste0(dir_busco_ref, ref, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/", busco)
        file_busco_fas <- paste0(dir_busco_gff, ".faa")
        file_busco_gff <- paste0(dir_busco_gff, ".gff")

        # output files
        file_busco_cds <- paste0(dir_busco_out, busco, ".cds.faa")
        file_busco_gff_out <- paste0(dir_busco_out, busco, ".gff")
        file_busco_bed_out <- paste0(dir_busco_out, busco, ".bed")
        file_busco_concat <- paste0(dir_busco_out, busco, ".faa")

        # extract header
        busco_header <- system(paste("grep '^>'", file_busco_fas), intern=T)

        # extract BUSCO coordinates
        coordinates <- f_extract_coordinates(busco_header, busco, ref)
        if (!is.null(coordinates$errmsg)) {
            f_write_log(fn_log=fn_log, msg=coordinates$errmsg)
            return(data.table::data.table(ref=ref, busco=busco, is_error=TRUE, identity=NA))
        }

        # manipulate GFF file
        if (!file.exists(file_busco_gff)) {
            f_create_gff(coordinates, busco, file_busco_gff_out)
            log4r::warn(fn_logger, paste0("File not found: ", busco, " GFF file for ", ref, ". Extracting the whole alignment."))
        } else {
            msg <- f_manipulate_gff(file_busco_gff, coordinates, busco, ref, file_busco_gff_out)
            if (!is.null(msg)) {
                if (!is.null(msg$errmsg)) {
                    f_write_log(fn_log=fn_log, msg=msg$errmsg)
                    return(data.table::data.table(ref=ref, busco=busco, is_error=TRUE, identity=NA))
                } else if (!is.null(msg$warnmsg)) {
                    f_write_log(fn_log=fn_log, msg=msg$warnmsg)
                }
            }
        }

        # run bedops gff2bed
        cmd_gff2bed <- paste(params$exe_gff2bed, "<", file_busco_gff_out, ">", file_busco_bed_out)
        system(cmd_gff2bed)
        
        # extract BUSCO
        f_extract_fasta_from_gff(file_fas, file_busco_gff_out, file_busco_cds, file_busco_concat, params$exe_gffread)

        # sanity check if FASTA is the same
        output <- f_compare_fasta(file_busco_fas, file_busco_concat)
        if (!is.null(output$errmsg)) {
            f_write_log(fn_log=fn_log, paste0("Warn: ", busco, " FASTA files for ", ref, ": ", output$errmsg, "."))
            return(data.table::data.table(ref=ref, busco=busco, is_error=TRUE, identity=NA))
        } else {
            if (!output$is_identical){
                score <- round(as.numeric(output$score),3)

                f_write_log(fn_log=fn_log, paste0("Warn: ", busco, " FASTA file for ", ref, ". Sequences are non-identical (identity=", score, ")."))

                if (score < min_ref_busco_identity) {
                    return(data.table::data.table(ref=ref, busco=busco, is_error=TRUE, identity=score))
                }
            }
        }

        return(data.table::data.table(ref=ref, busco=busco, is_error=FALSE, identity=100))
    }

    # update the list of output data.table
    df_chr_busco <- rbind(df_chr_busco, df_output_temp)

    log4r::info(fn_logger, paste0("File created/modified: BUSCO FASTA alignments for ", ref, "."))
}

stopCluster(nwcl)

# save data.table
data.table::fwrite(df_chr_busco, file=fn_refs_busco, quote=F, sep="\t")

# exclude problematic BUSCOs from further analyses
ls_problematic_busco <- unique(df_chr_busco$busco[df_chr_busco$is_error])
shared_busco <- shared_busco[!shared_busco %in% ls_problematic_busco]

# print the number of BUSCO in HTML document
cat(paste("Number of shared BUSCO after filtering non-identical genes:", length(shared_busco)))
f_write_log(fn_log=fn_log, msg=c("", paste0("Number of BUSCO after filtering problematic genes: ", length(shared_busco), ".")))

# stop the knit if shared busco is zero
if (length(shared_busco) == 0) {
    knitr::knit_exit()
}
```

## Extract BUSCOs for mapped reads
```{r results='asis'}
# open metadata file
df_mapped_reads_busco <- NULL
if (file.exists(fn_mapped_reads_busco)) {
    df_mapped_reads_busco <- data.table::fread(fn_mapped_reads_busco)
}

# filter out BUSCOs according to several criteria
if (!is.null(df_mapped_reads_busco) && !params$redo) {
    if (!params$include_incomplete) {
        # remove BUSCOs with incomplete taxa
        df_mapped_reads_busco <- df_mapped_reads_busco %>% mutate(prefix=paste0(read,"--",ref))
        df_mapped_reads_busco <- df_mapped_reads_busco %>%
                                    group_by(busco) %>%
                                    filter(n_distinct(prefix) == n_distinct(df_mapped_reads_busco$prefix))
        
        # update list
        shared_busco <- shared_busco[shared_busco %in% unique(df_mapped_reads_busco$busco)]

        # remove BUSCOs with low coverage
        ls_problematic_busco <- unique(df_mapped_reads_busco$busco[df_mapped_reads_busco$is_error])
        shared_busco <- shared_busco[!shared_busco %in% ls_problematic_busco]
    } else {
        # count the number of errors per BUSCO
        df_mapped_reads_busco_count <- df_mapped_reads_busco %>%
            group_by(busco,read) %>%
            summarise(n_ref=sum(ifelse(!is_error, 1, 0))) %>%
            group_by(busco) %>%
            summarise(n_read=sum(ifelse(n_ref>0, 1, 0)))
        
        # update the BUSCO list
        ls_problematic_busco <- unique(df_mapped_reads_busco_count$busco[df_mapped_reads_busco_count$n_read < params$min_taxa])
        shared_busco <- shared_busco[!shared_busco %in% ls_problematic_busco]
    }
} else {
    # run BUSCO extraction for mapped reads
    if (params$type == "coordinate" || params$type == "" || is.null(params$type)) {
        knitr::knit_child(paste0(params$codedir,"/2_busco_check/2a_coordinate.Rmd"))
    } else if (params$type == "pipeline") {
        knitr::knit_child(paste0(params$codedir,"/2_busco_check/2b_pipeline.Rmd"))
    }
}

# print the number of BUSCO in HTML document
cat(paste("Number of shared BUSCO after filtering low-coverage and incomplete genes:", length(shared_busco)))
f_write_log(fn_log=fn_log, msg=c("", paste0("Number of BUSCO after filtering problematic genes: ", length(shared_busco), ".")))
```

## Convert individual BUSCO alignments into MSA
```{r check-busco-msa}
# set up variables
busco_refs_dir_mid <- "/fasta/"
busco_read_dir_mid <- "/"
if (params$type == "pipeline") {
    busco_refs_dir_mid <- paste0("/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/")
    busco_read_dir_mid <- paste0("/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/")
}

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over shared BUSCOs
output <- foreach (busco = shared_busco, .combine='c') %dopar% {
    dir_output <- paste0(dir_busco_tree, busco, "/")
    if (!dir.exists(dir_output)) {
        dir.create(dir_output, recursive=T)
    }

    # output MSA file
    fn_out <- paste0(dir_output, busco, ".fa")
    fn_out_aligned <- paste0(dir_output, busco, "_aligned.fa")
    if (all(file.exists(fn_out, fn_out_aligned)) && !params$redo) {
        # check if file is not empty
        if (file.size(fn_out_aligned) != 0L) {
            return(NULL)
        }
    }

    # extract BUSCO alignment from reference sequences
    for (ref in ls_refseq) {
        file_busco_ref <- paste0(dir_busco_ref, ref, busco_refs_dir_mid, busco, ".fna")
        if (!file.exists(file_busco_ref)) {
            next
        }

        f_fasta2msa(file_busco_ref, ref, fn_out)
    }

    # extract BUSCO alignment from short reads
    count_unavailable_prefix <- 0
    for (read in ls_prefix) {
        file_busco_read <- paste0(dir_busco_read, read, busco_read_dir_mid, busco, ".fna")
        if (!file.exists(file_busco_read)) {
            count_unavailable_prefix <- count_unavailable_prefix + 1
            next
        }

        f_fasta2msa(file_busco_read, read, fn_out)
    }

    # align using MAFFT FFT-NS-2
    f_mafft(fn_out, fn_out_aligned, "--retree 2", params$exe_mafft)

    # return BUSCOs with zero mapped reads
    if (count_unavailable_prefix == length(ls_prefix)) {
        return(busco)
    }
}

stopCluster(nwcl)

# update the number of shared BUSCOs
shared_busco <- shared_busco[!shared_busco %in% output]

# print the number of BUSCO in HTML document
cat(paste("Number of shared BUSCO after filtering reference-only genes:", length(shared_busco)))
f_write_log(fn_log=fn_log, msg=c("", paste0("Number of BUSCO after filtering genes with zero mapped reads: ", length(shared_busco), ".")))

# stop the knit if shared busco is zero
if (length(shared_busco) == 0) {
    knitr::knit_exit()
}
```

## Generate BUSCO trees
```{r check-busco-tree, echo=FALSE}
f_write_log(fn_log=fn_log, msg=c("", "------------ BUSCO Trees -----------"))
              
# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over shared BUSCOs
foreach (busco = shared_busco) %dopar% {
    dir_output <- paste0(dir_busco_tree, busco, "/")

    # output IQ-Tree2 files
    fn_fasta <- paste0(dir_output, busco, "_aligned.fa")
    fn_treefile <- paste0(dir_output, busco, "_aligned.fa.treefile")
    if (!file.exists(fn_treefile) || params$redo) {
        # run IQ-Tree2
        f_iqtree2(fn_fasta, params$exe_iqtree2)
    }

    # change the ID into species name
    fn_treefile_species <- paste0(dir_output, busco, "_aligned.fa.treefile.species")
    if (!file.exists(fn_treefile_species) || params$redo) {
        f_tips_to_species(fn_treefile, fn_treefile_species, ls_species_name)
    }
}

stopCluster(nwcl)
```

```{r check-busco-treeshrink, echo=FALSE}
f_write_log(fn_log=fn_log, msg=c("", "------------ TreeShrink ------------"))
              
# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over shared BUSCOs
foreach (busco = shared_busco) %dopar% {
    dir_output <- paste0(dir_busco_tree, busco, "/treeshrink/")

    # input file
    fn_treefile <- paste0(dir_busco_tree, busco, "/", busco, "_aligned.fa.treefile")
    if (!file.exists(fn_treefile)) {
        return(NULL)
    }

    # output file
    prefix <- paste0(busco, ".treeshrink")

    fn_treeshrink_tree <- paste0(dir_output, prefix, ".treefile")
    if (!file.exists(fn_treeshrink_tree) || params$redo) {
        # run TreeShrink
        f_treeshrink(fn_treefile, prefix, dir_output, params$exe_treeshrink)
    }

    # change the ID into species name
    fn_treefile_species <- paste0(dir_output, prefix, ".treefile.species")
    if (!file.exists(fn_treefile_species) || params$redo) {
        f_tips_to_species(fn_treeshrink_tree, fn_treefile_species, ls_species_name)
    }
}

log4r::info(fn_logger, "File created/modified: BUSCO trees.")
stopCluster(nwcl)
```