---
title: "Read Mapping of Short Reads on Reference Sequences"
---

```{r check-readmap-setup, include=FALSE}
f_write_log(fn_log=fn_log,
            msg=c("", "####################################",
                      "####        Read Mapping        ####",
                      "####################################"))

# refseq and short reads directories
dir_refseq <- paste0(getwd(), "/refseq/")
dir_shortreads <- paste0(getwd(), "/short_reads/")

# list of refseq
ls_refseq <- list.dirs(dir_refseq, recursive=F, full.names=F)

# create output directory
dir_shortreads <- paste0(getwd(), "/busco_check/short_reads/")
dir_readmap <- paste0(getwd(), "/busco_check/readmap/")

# create directory if not exist
lapply(list(dir_shortreads,dir_readmap), function(x){if(!dir.exists(x)) dir.create(x, recursive=T)})

# set thread for bwamem2 and samtools
thread_bwamem2 <- ifelse(params$thread < 50, params$thread, 50)
thread_adapterremoval <- ifelse(params$thread < 10, params$thread, 10)
thread_samtools <- ifelse(params$thread < 10, params$thread, 10)
```

## QC of short reads
```{check-qc-reads}
# select short reads
df_shortreads <- data.table::fread(params$file_shortreads)
ls_shortreads <- df_shortreads$id[df_shortreads$is_check]

# create doSNOW cluster
nwcl <- makeCluster(floor(params$thread/thread_adapterremoval))
doSNOW::registerDoSNOW(nwcl)

# iterate over short reads
foreach (read = ls_shortreads) %dopar% {
    prefix <- paste0(dir_shortreads, read)

    # check if file exists
    file_fastq <- paste0(prefix, ".collapsed.truncated")
    if (file.exists(file_fastq) && !params$redo) {
        log4r::warn(fn_logger, paste0("File found: filtered FASTQ file for ", prefix, ". Skipped."))
        return(NULL)
    }

    # extract FASTQ files
    dir_reads <- paste0(dir_shortreads, read, "/fastq/")
    ls_fastq <- list.files(dir_reads, pattern="*.fastq", recursive=F, full.names=T)
    len_ls_fastq <- length(ls_fastq)

    # run BWA-MEM2
    if (len_ls_fastq != 1 && len_ls_fastq != 2) {
        log4r::warn(fn_logger, paste0("Warn: invalid number of FASTQ file for ", prefix, " (", len_ls_fastq, "). Skipped."))
        return(NULL)
    }
    
    f_qc_short_reads(ls_fastq, params$file_adapters, prefix, 25, thread_adapterremoval, params$exe_adapterremoval)
    log4r::info(fn_logger, paste0("File created: filtered FASTQ file for ", prefix, "."))
}

stopCluster(nwcl)
```

## Map short reads to reference using BWA-MEM2
```{r check-readmap}
# create doSNOW cluster
nwcl <- makeCluster(floor(params$thread/thread_bwamem2))
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
for (ref in ls_refseq) {
    # extract reference sequence
    dir_ref <- paste0(dir_refseq, ref, "/ncbi_dataset/data/", ref, "/")
    file_ref <- list.files(dir_ref, pattern="*.fna$", recursive=F, full.names=T)
    
    if (length(file_ref) != 1) {
        file_ref <- file_ref[1]
        log4r::warn(fn_logger, paste0("Multiple references are found for ", ref, ". ", file_ref, " is selected."))
    }

    # iterate over short reads
    foreach (read = ls_shortreads) %dopar% {
        prefix <- paste0(read, "--", ref)

        # output directory
        dir_output <- paste0(dir_readmap, prefix, "/")
        if (!dir.exists(dir_output)) {
            dir.create(dir_output, recursive=T)
        }

        # check if file exists
        file_sam <- paste0(dir_output, prefix, ".sam")
        if (file.exists(file_sam) && !params$redo) {
            log4r::warn(fn_logger, paste0("File found: SAM file for ", prefix, ". Skipped."))
            return(NULL)
        }

        # extract FASTQ file
        file_fastq <- paste0(dir_shortreads, read, ".collapsed.truncated")
        if (!file.exists(file_fastq)) {
            log4r::warn(fn_logger, paste0("File not found: filtered FASTQ file for ", read, ". Skipped."))
            return(NULL)
        }
        
        # run BWA-MEM2
        f_read_mapping(file_ref, file_fastq, thread_bwamem2, params$exe_bwamem2, file_sam)
        log4r::info(fn_logger, paste0("File created: SAM file for ", prefix, "."))
    }
}

stopCluster(nwcl)
```

## Convert SAM to BAM
```{r check-readmap-conversion}
f_write_log(fn_log=fn_log,
            msg=c("", "--------- Data Preprocessing -------"))

# create doSNOW cluster
nwcl <- makeCluster(floor(params$thread/thread_samtools))
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
for (ref in ls_refseq) {
    # extract reference sequence
    dir_ref <- paste0(dir_refseq, ref, "/ncbi_dataset/data/", ref, "/")
    file_ref <- list.files(dir_ref, pattern="*.fna$", recursive=F, full.names=T)
    
    if (length(file_ref) != 1) {
        file_ref <- file_ref[1]
        log4r::warn(fn_logger, paste0("Multiple references are found for ", ref, ". ", file_ref, " is selected."))
    }

    # iterate over short reads
    foreach (read = ls_shortreads) %dopar% {
        prefix <- paste0(read, "--", ref)
        
        # output directory
        dir_output <- paste0(dir_readmap, prefix, "/")

        # check if file exists
        file_bam <- paste0(dir_output, prefix, ".bam")
        file_vcf <- paste0(dir_output, prefix, ".vcf.gz")
        file_vcf_idx <- paste0(file_vcf, ".tbi")
        if (all(file.exists(file_bam, file_vcf, file_vcf_idx)) && !params$redo) {
            log4r::warn(fn_logger, paste0("File found: BAM and VCF files for ", prefix, ". Skipped."))
            return(NULL)
        }

        # convert SAM to BAM and VCF
        f_variant_calling(prefix, dir_output, thread_samtools, file_ref, params$exe_samtools, params$exe_bcftools)
        log4r::info(fn_logger, paste0("File created: BAM and VCF files for ", prefix, "."))
    }
}

stopCluster(nwcl)
```