---
title: "Run Coordinate-based BUSCO Analysis on Reference Sequences and Mapped Reads"
---

## Extract BUSCO based on reference coordinates
```{r check-busco-coordinate-extract}
f_write_log(fn_log=fn_log, msg=c("", "------------ Mapped Reads ----------"))

# output data.table
df_mapped_reads_busco <- data.table::data.table(ref=character(), read=character(), busco=character(), is_error=logical(), note=character())

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# store the list of problematic BUSCOs
ls_error_busco <- c()

# iterate over short reads
for (prefix in ls_prefix) {
    # create outdir
    dir_busco_out <- paste0(dir_busco_read, prefix, "/")
    if (!dir.exists(dir_busco_out)) {
        dir.create(dir_busco_out, recursive=T)
    }

    # output file
    fn_read_summary <- paste0(dir_busco_out, "metadata.tsv")
    if (file.exists(fn_read_summary) && !params$redo) {
        # update the data.frame
        df_read_summary <- data.table::fread(fn_read_summary)
        df_mapped_reads_busco <- rbind(df_mapped_reads_busco, df_read_summary)

        ls_error_busco <- c(ls_error_busco, df_mapped_reads_busco$busco[df_mapped_reads_busco$is_error])
        ls_error_busco <- unique(ls_error_busco)

        next
    }


    # extract short read and reference sequence IDs
    read <- unlist(strsplit(prefix, split="--"))[1]
    ref <- unlist(strsplit(prefix, split="--"))[2]

    # check input files
    file_fas <- paste0(dir_readmap, prefix, "/", prefix, ".fa")
    if (!file.exists(file_fas)) {
        log4r::warn(fn_logger, paste0("File not found: FASTA file for ", prefix, ". Skipped."))
        next
    }

    # iterate over shared BUSCOs
    df_output_temp <- foreach (busco = shared_busco, .combine='rbind') %dopar% {
        require(Biostrings)
        require(stringr)

        # input files
        file_bam <- paste0(dir_readmap, prefix, "/", prefix, ".bam")
        file_busco_gff_out <- paste0(dir_busco_ref, ref, "/fasta/", busco, ".gff")
        file_busco_bed_out <- paste0(dir_busco_ref, ref, "/fasta/", busco, ".bed")
        if (!all(file.exists(file_busco_gff_out, file_busco_bed_out))) {
            log4r::warn(fn_logger, msg=paste0("File not found: ", busco, " GFF or BED for ", ref, ". Skipped."))
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=TRUE, note="File not found: GFF or BED files"))
        }

        # output files
        file_busco_cds <- paste0(dir_busco_out, busco, ".cds.faa")
        file_busco_concat <- paste0(dir_busco_out, busco, ".faa")

        # check if file exists
        if (file.exists(file_busco_concat) && file.size(file_busco_concat) != 0L && !params$redo) {
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=FALSE, note="File found: FASTA sequence file"))
        }

        # check if BUSCO is error for other mapped reads
        if (busco %in% ls_error_busco) {
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=TRUE, note="BUSCO is problematic for other mapped reads"))
        }

        # calculate read depth
        read_coverage <- f_calculate_read_coverage(file_bam, file_busco_bed_out, params$exe_samtools)
        if (is.null(read_coverage) || is.na(read_coverage)) {
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=TRUE, note="Error: invalid coverage"))
        } else if (read_coverage >= params$min_busco_depth && read_coverage <= params$max_busco_depth) {
            f_extract_fasta_from_gff(file_fas, file_busco_gff_out, file_busco_cds, file_busco_concat, params$exe_gffread)
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=FALSE, note=NA))
        } else if (read_coverage < params$min_busco_depth || read_coverage > params$max_busco_depth) {
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=TRUE, note=paste("Average CDS coverage:", read_coverage)))
        }
    }

    # update data.table
    df_mapped_reads_busco <- rbind(df_mapped_reads_busco, df_output_temp)
    ls_error_busco <- c(ls_error_busco, df_mapped_reads_busco$busco[df_mapped_reads_busco$is_error])
    ls_error_busco <- unique(ls_error_busco)

    data.table::fwrite(df_output_temp, file=fn_read_summary, sep="\t", quote=F, row.names=F)
    log4r::info(fn_logger, paste0("File created/modified: BUSCO FASTA alignments for ", prefix, "."))
}

stopCluster(nwcl)

# save the output file
data.table::fwrite(df_mapped_reads_busco, file=fn_mapped_reads_busco, sep="\t", quote=F)

# remove BUSCOs with incomplete taxa
df_mapped_reads_busco <- df_mapped_reads_busco %>% mutate(prefix=paste0(read,"--",ref))
df_mapped_reads_busco <- df_mapped_reads_busco %>%
                            group_by(busco) %>%
                            filter(n_distinct(prefix) == n_distinct(df_mapped_reads_busco$prefix))

# update list
shared_busco <- shared_busco[shared_busco %in% unique(df_mapped_reads_busco$busco)]

# remove BUSCOs with low coverage
ls_problematic_busco <- unique(df_mapped_reads_busco$busco[df_mapped_reads_busco$is_error])
shared_busco <- shared_busco[!shared_busco %in% ls_problematic_busco]~

# stop the knit if shared busco is zero
if (length(shared_busco) == 0) {
    knitr::knit_exit()
}
```

```{r check-busco-coordinate-coverage}
# check coverage for shared BUSCO
for (ref in ls_refseq) {
    # output file
    fn_busco_gff_all <- paste0(dir_busco_ref, ref, "/all_busco.gff")

    # concatenate all GFF files
    shared_busco_gff <- paste0(dir_busco_ref, ref, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/", shared_busco, ".gff")
    shared_busco_gff <- paste0(shared_busco_gff, collapse=" ")
    system(paste("cat", shared_busco_gff, ">", fn_busco_gff_all))
}

# iterate over mapped reads
df_coverage <- data.table::data.table()

for (prefix in ls_prefix) {
    # extract reference and reads
    readref <- strsplit(prefix, split="--")[[1]]

    # input parameters
    dir_qualimap_output <- paste0(dir_busco_read, prefix, "/qualimap/")
    file_bam <- paste0(dir_readmap, prefix, "/", prefix, ".bam")
    file_gff <- paste0(dir_busco_ref, readref[2], "/all_busco.gff")

    # run QualiMap
    file_qualimap_html <- paste0(dir_busco_read, prefix, "/qualimap/qualimapReport.html")
    if (!file.exists(file_qualimap_html)) {
        f_qualimap(file_bam, dir_qualimap_output, params$thread, file_gff, params$exe_qualimap)
    }

    # extract coverage
    mean_coverage <- system(paste0("grep 'mean coverageData' ", dir_qualimap_output, "genome_results.txt"), intern=T)
    mean_coverage <- strsplit(mean_coverage, split=" = ")[[1]][2]
    mean_coverage <- as.numeric(strsplit(mean_coverage, split="X")[[1]][1])

    # extract percentage of mapped reads
    mapped_reads <- system(paste0("grep 'number of mapped reads' ", dir_qualimap_output, "genome_results.txt"), intern=T)
    mapped_reads <- stringr::str_match(mapped_reads, "\\((.*)\\)")[,2]
    mapped_reads <- as.numeric(strsplit(mapped_reads, split="%")[[1]][1])

    # update the data.table
    df_coverage <- rbind(df_coverage, data.table::data.table(read_id=readref[1], ref_id=readref[2],
                                                             read_sp=df_reads$species[df_reads$id==readref[1]],
                                                             ref_sp=df_refs$species[df_refs$id==readref[2]],
                                                             mean_coverage=mean_coverage,
                                                             mapped_reads=mapped_reads))
}

# save the file
fn_output <- paste0(dir_busco_read, "summary_coverage.tsv")
data.table::fwrite(df_coverage, file=fn_output, sep="\t", quote=F, row.names=F, col.names=F)
```
