---
title: "Run BUSCO Pipeline on Individual Reference Sequences and Mapped Reads"
---

## Run BUSCO on mapped reads
```{r, include=FALSE}
f_write_log(fn_log=fn_log, msg=c("", "------------ Mapped Reads ----------"))

# set up variables
is_busco_read <- FALSE

ls_busco_read_done <- c()
ls_busco_read_run <- c()

# iterate over mapped sequences
if (params$redo) {
    ls_busco_read_run <- ls_prefix
} else {
    for (read in ls_prefix) {
        # check if BUSCO run exists
        file_busco_log <- paste0(dir_busco_read, read, "/logs/busco.log")
        if (file.exists(file_busco_log)) {
            finish_msg <- system(paste("grep 'BUSCO analysis done'", file_busco_log), intern=T)
            if (length(finish_msg) != 0) {
                ls_busco_read_done <- c(ls_busco_read_done, paste("-", read))
                next
            }
        }

        # add mapped sequence to to-do-list
        ls_busco_read_run <- c(ls_busco_read_run, read)
    }
}

# output log file for available mapped sequence BUSCO runs
if (length(ls_busco_read_done) > 0) {
    ls_busco_read_done <- c(paste0("Available mapped sequence BUSCO runs (", length(ls_busco_read_done), "/", length(ls_prefix), ")"), ls_busco_read_done, "")
    f_write_log(fn_log=fn_log, msg=ls_busco_read_done)
}

# update variable for the next analysis
if (length(ls_busco_read_run) > 0) {
    is_busco_read <- TRUE
}
```

```{r}
# print the number of available mapped reads BUSCO in HTML document
cat(paste0("Available mapped reads BUSCO (", length(ls_busco_read_done), "/", length(ls_prefix), ")"))
cat(paste("Mapped reads BUSCO to run:", length(ls_busco_read_run)))
```

```{r check-busco-pipeline-run, include=is_busco_read, eval=is_busco_read}
f_write_log(fn_log=fn_log, msg=paste("Running BUSCO using", params$thread_busco, "threads"))

# create doSNOW cluster
cl_busco <- floor(params$thread/params$thread_busco)
if (cl_busco < 1) {
    log4r::error(fn_logger, "Invalid thread for BUSCO. Exited.")
    knitr::knit_exit()
}

nwcl <- makeCluster(cl_busco)
doSNOW::registerDoSNOW(nwcl)

# iterate over mapped reads
foreach (prefix = ls_busco_read_run) %dopar% {
    # input FASTA file
    file_read <- paste0(dir_readmap, prefix, "/", prefix, ".fa")

    # run BUSCO
    f_run_busco(file_read, dir_lineage, prefix, dir_busco_read, params$busco_mode, params$thread_busco, params$exe_busco)
    log4r::info(fn_logger, paste0("File created: BUSCO run for ", prefix, "."))
}

stopCluster(nwcl)
```

## Extract common BUSCOs between reference sequences and mapped reads
```{r check-busco-pipeline-extract}
# output data.table
df_mapped_reads_busco <- data.table::data.table(ref=character(), read=character(), busco=character(), is_error=logical(), note=character())

# store the list of problematic BUSCOs
ls_error_busco <- c()

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over mapped reads
for (prefix in ls_prefix) {
    # input files
    file_bam <- paste0(dir_readmap, prefix, "/", prefix, ".bam")

    # output directories
    dir_bed <- paste0(dir_busco_read, prefix, "/bed/")
    if (!dir.exists(dir_bed)) {
        dir.create(dir_bed, recursive=T)
    }

    # output file
    fn_read_summary <- paste0(dir_busco_read, prefix, "/metadata.tsv")
    if (file.exists(fn_read_summary) && !params$redo) {
        # update the data.frame
        df_read_summary <- data.table::fread(fn_read_summary)
        df_mapped_reads_busco <- rbind(df_mapped_reads_busco, df_read_summary)

        ls_error_busco <- c(ls_error_busco, df_mapped_reads_busco$busco[df_mapped_reads_busco$is_error])
        ls_error_busco <- unique(ls_error_busco)

        next
    }

    # directory for BUSCO alignments
    dir_buscos <- paste0(dir_busco_read, prefix, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/")
    
    # extract short read and reference sequence IDs
    read <- unlist(strsplit(prefix, split="--"))[1]
    ref <- unlist(strsplit(prefix, split="--"))[2]

    # get average coverage
    mean_coverage <- as.numeric(df_readmap_coverage$mean_coverage[df_readmap_coverage$read_id==read & df_readmap_coverage$ref_id==ref])
    max_busco_depth <- 2 * mean_coverage

    # remove file extension .faa
    busco_ids <- list.files(dir_buscos, pattern = "*.faa$", full.names = F, recursive = F)
    busco_ids <- sapply(busco_ids, function(x) { gsub(".faa", "", x) })
    
    # iterate over shared BUSCOs
    df_output_temp <- foreach (busco = busco_ids, .combine='rbind') %dopar% {
        # input files
        file_busco_gff <- paste0(dir_buscos, busco, ".gff")
        if (!file.exists(file_busco_gff)) {
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=TRUE, note="File not found: GFF file"))
        }

        # check if BUSCO is error for other mapped reads
        if (busco %in% ls_error_busco) {
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=TRUE, note="BUSCO is problematic for other mapped reads"))
        }

        # run bedops gff2bed
        file_busco_bed <- paste0(dir_bed, busco, ".bed")
        cmd_gff2bed <- paste(params$exe_gff2bed, "<", file_busco_gff, ">", file_busco_bed)
        system(cmd_gff2bed)

        # calculate read depth
        read_coverage <- f_calculate_read_coverage(file_bam, file_busco_bed, params$exe_samtools)
        if (is.null(read_coverage) || is.na(read_coverage)) {
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=TRUE, note="Error: invalid coverage"))
        } else if (read_coverage >= params$min_busco_depth && read_coverage <= max_busco_depth) {
            f_extract_fasta_from_gff(file_fas, file_busco_gff_out, file_busco_cds, file_busco_concat, params$exe_gffread)
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=FALSE, note=NA))
        } else if (read_coverage < params$min_busco_depth || read_coverage > max_busco_depth) {
            return(data.table::data.table(ref=ref, read=read, busco=busco, is_error=TRUE, note=paste("Average CDS coverage:", read_coverage)))
        }
    }

    # update data.table
    df_mapped_reads_busco <- rbind(df_mapped_reads_busco, df_output_temp)
    ls_error_busco <- c(ls_error_busco, df_mapped_reads_busco$busco[df_mapped_reads_busco$is_error])
    ls_error_busco <- unique(ls_error_busco)

    data.table::fwrite(df_output_temp, file=fn_read_summary, sep="\t", quote=F, row.names=F)
    log4r::info(fn_logger, paste0("File created/modified: BUSCO FASTA alignments for ", prefix, "."))
}

stopCluster(nwcl)

# save the output file
data.table::fwrite(df_mapped_reads_busco, file=fn_mapped_reads_busco, sep="\t", quote=F)

# remove BUSCOs with incomplete taxa
df_mapped_reads_busco <- df_mapped_reads_busco %>% mutate(prefix=paste0(read,"--",ref))
df_mapped_reads_busco <- df_mapped_reads_busco %>%
                            group_by(busco) %>%
                            filter(n_distinct(prefix) == n_distinct(df_mapped_reads_busco$prefix))

# update list
shared_busco <- shared_busco[shared_busco %in% unique(df_mapped_reads_busco$busco)]

# remove BUSCOs with low coverage
ls_problematic_busco <- unique(df_mapped_reads_busco$busco[df_mapped_reads_busco$is_error])
shared_busco <- shared_busco[!shared_busco %in% ls_problematic_busco]

# check if shared BUSCOs found
if (length(shared_busco) == 0) {
    log4r::error(fn_logger, paste0("Error: no BUSCO shared between sequences. Exited."))
    knitr::knit_exit()
}
```

```{r check-busco-pipeline-coverage}
# iterate over mapped reads
df_coverage <- data.table::data.table()

for (prefix in ls_prefix) {
    # extract reference and reads
    readref <- strsplit(prefix, split="--")[[1]]

    # input parameters
    dir_qualimap_output <- paste0(dir_busco_read, prefix, "/qualimap/")
    file_bam <- paste0(dir_readmap, prefix, "/", prefix, ".bam")
    file_gff <- paste0(dir_busco_read, prefix, "/all_busco.gff")

    # concatenate all GFF files
    shared_busco_gff <- paste0(dir_busco_read, prefix, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/", shared_busco, ".gff")
    shared_busco_gff <- paste0(shared_busco_gff, collapse=" ")
    system(paste("cat", shared_busco_gff, ">", file_gff))    

    # run QualiMap
    file_qualimap_html <- paste0(dir_busco_read, prefix, "/qualimap/qualimapReport.html")
    if (!file.exists(file_qualimap_html)) {
        f_qualimap(file_bam, dir_qualimap_output, params$thread, file_gff, params$exe_qualimap)
    }

    # extract coverage
    mean_coverage <- system(paste0("grep 'mean coverageData' ", dir_qualimap_output, "genome_results.txt"), intern=T)
    mean_coverage <- strsplit(mean_coverage, split=" = ")[[1]][2]
    mean_coverage <- as.numeric(strsplit(mean_coverage, split="X")[[1]][1])

    # extract percentage of mapped reads
    mapped_reads <- system(paste0("grep 'number of mapped reads' ", dir_qualimap_output, "genome_results.txt"), intern=T)
    mapped_reads <- stringr::str_match(mapped_reads, "\\((.*)\\)")[,2]
    mapped_reads <- as.numeric(strsplit(mapped_reads, split="%")[[1]][1])

    # update the data.table
    df_coverage <- rbind(df_coverage, data.table::data.table(read_id=readref[1], ref_id=readref[2],
                                                             read_sp=df_reads$species[df_reads$id==readref[1]],
                                                             ref_sp=df_refs$species[df_refs$id==readref[2]],
                                                             mean_coverage=mean_coverage,
                                                             mapped_reads=mapped_reads))
}

# save the file
fn_output <- paste0(dir_busco_read, "summary_coverage.tsv")
data.table::fwrite(df_coverage, file=fn_output, sep="\t", quote=F, row.names=F, col.names=F)
```
