---
title: "Run BUSCO Pipeline on Reference Sequences and Short Reads"
---

```{r check-busco-setup, include=FALSE}
f_write_log(fn_log=fn_log,
            msg=c("", "####################################",
                      "####            BUSCO           ####",
                      "####################################"))

# list of reference sequences
dir_refseq <- paste0(getwd(), "/refseq/")
ls_refseq <- list.dirs(dir_refseq, recursive=F, full.names=F)

# output directories
dir_readmap <- paste0(getwd(), "/busco_check/readmap/")

# create output directory
dir_busco <- paste0(getwd(), "/busco_check/busco/")
if (!dir.exists(dir_busco)) {
  dir.create(dir_busco, recursive = T)
}

# output file
fn_summary <- paste0(dir_busco, params$prefix, ".sumtable")

# set thread for BUSCO
thread_busco <- ifelse(params$thread < 50, params$thread, 50)
```

## Download BUSCO lineage dataset
```{r busco-lineage-download}
# create directory for BUSCO lineage
dir_busco_lineage <- paste0(getwd(), "/busco_lineage/")
if (!dir.exists(dir_busco_lineage)) {
    dir.create(dir_busco_lineage, recursive=T)
}

# specific BUSCO lineage
dir_lineage <- paste0(dir_busco_lineage, "lineages/", params$busco_lineage)
if (!dir.exists(dir_lineage)) {
    busco_cmd <- paste(params$exe_busco, "--download_path", dir_busco_lineage, "--download", params$busco_lineage)
    system(busco_cmd)
}
```

## Run BUSCO on reference sequences
```{r check-busco-ref}
f_write_log(fn_log=fn_log, msg=c("", "--------- Reference Sequence -------"))

dir_busco_ref <- paste0(dir_busco, "refseq/")
if (!dir.exists(dir_busco_ref)) {
    dir.create(dir_busco_ref, recursive=T)
}

# create doSNOW cluster
nwcl <- makeCluster(floor(params$thread/thread_busco))
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
foreach (ref = ls_refseq) %dopar% {
    # initiate variable
    is_busco_run <- TRUE

    # extract reference sequence
    dir_ref <- paste0(dir_refseq, ref, "/ncbi_dataset/data/", ref, "/")
    file_ref <- list.files(dir_ref, pattern="*.fna$", recursive=F, full.names=T)
    
    if (length(file_ref) > 1) {
        file_ref <- file_ref[1]
        log4r::warn(fn_logger, paste0("Multiple references are found for ", ref, ". ", file_ref, " is selected."))
    }

    # check if BUSCO run exists
    file_busco_log <- paste0(dir_busco_ref, ref, "/logs/busco.log")
    if (file.exists(file_busco_log) && !params$redo) {
        finish_msg <- system(paste("grep 'BUSCO analysis done'", file_busco_log), intern=T)
        if (length(finish_msg) != 0) {
            log4r::info(fn_logger, paste0("File found: BUSCO run for ", ref, ". Skipped."))
            is_busco_run <- FALSE
        }
    }

    # run BUSCO
    if (is_busco_run) {
        f_run_busco(file_ref, dir_lineage, ref, dir_busco_ref, params$busco_mode, thread_busco, params$exe_busco)
        log4r::info(fn_logger, paste0("File created: BUSCO run for ", ref, "."))
    }

    return(NULL)
}

stopCluster(nwcl)
```

## Extract common BUSCOs between reference sequences
```{r check-busco-common}
# list BUSCO directories for reference sequences
ls_refseq <- list.dirs(dir_busco_ref, full.names=F, recursive=F)

# initiate variable
shared_busco <- c()

# iterate over reference sequences
for (ref in ls_refseq) {
  # directory for BUSCO alignments
  dir_buscos <- paste0(dir_busco_ref, ref, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/")
  
  # remove file extension .fna
  busco_ids <- list.files(dir_buscos, pattern = "*.fna$", full.names = F, recursive = F)
  busco_ids <- sapply(busco_ids, function(x) { gsub(".fna", "", x) })
  
  # extract shared BUSCOs between reference sequences
  if (length(shared_busco) == 0) {
    shared_busco <- busco_ids
  } else {
    shared_busco <- intersect(shared_busco, busco_ids)
  }
}

# check if shared BUSCOs found
if (length(shared_busco) == 0) {
    log4r::error(fn_logger, paste0("Error: no BUSCO shared between reference sequences. Exited."))
    knitr::knit_exit()
} else {
    f_write_log(fn_log=fn_log, msg=c("", paste0("Number of complete single-copy BUSCOs: ", length(shared_busco), ".")))
}
```

## Extract BUSCOs for reference sequences
```{r check-busco-extract-ref}
# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
for (ref in ls_refseq) {
    # directory for BUSCO alignments
    dir_busco_out <- paste0(dir_busco_ref, ref, "/fasta/")
    if (!dir.exists(dir_busco_out)) {
        dir.create(dir_busco_out, recursive=T)
    }
  
    # input file
    dir_ref <- paste0(dir_refseq, ref, "/ncbi_dataset/data/", ref, "/")
    file_fas <- list.files(dir_ref, pattern="*.fna$", recursive=F, full.names=T)[1]

    # iterate over shared BUSCOs
    foreach (busco = shared_busco) %dopar% {
        require(data.table)
        require(Biostrings)
        require(stringr)

        # BUSCO files
        dir_busco_gff <- paste0(dir_busco_ref, ref, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/", busco)
        file_busco_fas <- paste0(dir_busco_gff, ".fna")
        file_busco_gff <- paste0(dir_busco_gff, ".gff")
        if (!file.exists(file_busco_fas)) {
            log4r::warn(fn_logger, paste0("File not found: FASTA file for ", ref, ". Skipped."))
            return(NULL)
        }
        
        # extract header
        busco_header <- system(paste("grep '^>'", file_busco_fas), intern=T)

        # output files
        file_busco_cds <- paste0(dir_busco_out, busco, ".cds.fna")
        file_busco_gff_out <- paste0(dir_busco_out, busco, ".gff")
        file_busco_concat <- paste0(dir_busco_out, busco, ".fna")

        # check if file exists
        if (file.exists(file_busco_concat) && !params$redo) {
            if (file.size(file_busco_concat) != 0L) {
                return(NULL)
            }
        }

        # extract BUSCO coordinates
        coordinates <- f_extract_coordinates(busco_header, busco, ref)
        if (!is.null(coordinates$errmsg)) {
            f_write_log(fn_log=fn_log, msg=coordinates$errmsg)
            return(NULL)
        }

        # manipulate GFF file
        if (!file.exists(file_busco_gff)) {
            f_create_gff(coordinates, busco, file_busco_gff_out)
            log4r::info(fn_logger, paste0("File created: ", busco, " GFF file for ", ref, "."))
        } else {
            msg <- f_manipulate_gff(file_busco_gff, coordinates, busco, ref, file_busco_gff_out)
            if (!is.null(msg$errmsg)) {
                f_write_log(fn_log=fn_log, msg=msg$errmsg)
                return(NULL)
            } else if (!is.null(msg$warnmsg)) {
                f_write_log(fn_log=fn_log, msg=msg$warnmsg)
            }
        }
        
        # extract BUSCO
        f_extract_fasta_from_gff(file_fas, file_busco_gff_out, file_busco_cds, file_busco_concat, params$exe_gffread)
    }

    log4r::info(fn_logger, paste0("File created/modified: BUSCOs for ", ref, "."))
}

stopCluster(nwcl)
```

## Extract BUSCOs for short reads
```{r check-busco-extract-reads}
f_write_log(fn_log=fn_log, msg=c("", "------------ Short Reads -----------"))

dir_busco_read <- paste0(dir_busco, "short_reads/")
if (!dir.exists(dir_busco_read)) {
    dir.create(dir_busco_read, recursive=T)
}

# extract list of short reads
ls_prefix <- list.dirs(dir_readmap, recursive=F, full.names=F)

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over short reads
for (prefix in ls_prefix) {
    # create outdir
    dir_busco_out <- paste0(dir_busco_read, prefix, "/")
    if (!dir.exists(dir_busco_out)) {
        dir.create(dir_busco_out, recursive=T)
    }

    # extract short read and reference sequence IDs
    ref <- unlist(strsplit(prefix, split="--"))[2]

    # check input files
    file_fas <- paste0(dir_readmap, prefix, "/", prefix, ".fa")
    if (!file.exists(file_fas)) {
        log4r::warn(fn_logger, paste0("File not found: FASTA file for ", prefix, ". Skipped."))
        next
    }

    # iterate over shared BUSCOs
    foreach (busco = shared_busco) %dopar% {
        require(Biostrings)
        require(stringr)

        # BUSCO file
        file_busco_gff_out <- paste0(dir_busco_ref, ref, "/fasta/", busco, ".gff")
        if (!file.exists(file_busco_gff_out)) {
            f_write_log(fn_log=fn_log, msg=paste0("File not found: ", busco, " GFF for ", ref, ". Skipped."))
            return(NULL)
        }

        # output files
        file_busco_cds <- paste0(dir_busco_out, busco, ".cds.fna")
        file_busco_concat <- paste0(dir_busco_out, busco, ".fna")

        # check if file exists
        if (file.exists(file_busco_concat) && !params$redo) {
            if (file.size(file_busco_concat) != 0L) {
                return(NULL)
            }
        }

        # extract BUSCO
        f_extract_fasta_from_gff(file_fas, file_busco_gff_out, file_busco_cds, file_busco_concat, params$exe_gffread)
    }

    log4r::info(fn_logger, paste0("File created/modified: BUSCOs for ", prefix, "."))
}

stopCluster(nwcl)
```

## Convert individual BUSCO alignments into MSA
```{r check-busco-msa}
dir_busco_tree <- paste0(dir_busco, "trees/")
if (!dir.exists(dir_busco_tree)) {
    dir.create(dir_busco_tree, recursive=T)
}

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over shared BUSCOs
foreach (busco = shared_busco) %dopar% {
    dir_output <- paste0(dir_busco_tree, busco, "/")
    if (!dir.exists(dir_output)) {
        dir.create(dir_output, recursive=T)
    }

    # output MSA file
    fn_out <- paste0(dir_output, busco, ".fa")
    fn_out_aligned <- paste0(dir_output, busco, "_aligned.fa")
    if (all(file.exists(fn_out, fn_out_aligned)) && !params$redo) {
        # check if file is not empty
        if (file.size(fn_out_aligned) != 0L) {
            return(NULL)
        }
    }

    # extract BUSCO alignment from reference sequences
    ls_busco_ref <- list.dirs(dir_busco_ref, recursive=F, full.names=F)
    for (ref in ls_busco_ref) {
        file_busco_ref <- paste0(dir_busco_ref, ref, "/fasta/", busco, ".fna")
        if (!file.exists(file_busco_ref)) {
            next
        }

        f_fasta2msa(file_busco_ref, ref, fn_out)
    }

    # extract BUSCO alignment from short reads
    ls_busco_reads <- list.dirs(dir_busco_read, recursive=F, full.names=F)
    for (read in ls_busco_reads) {
        file_busco_read <- paste0(dir_busco_read, read, "/", busco, ".fna")
        if (!file.exists(file_busco_read)) {
            next
        }

        f_fasta2msa(file_busco_read, read, fn_out)
    }

    # align using MAFFT FFT-NS-2
    f_mafft(fn_out, fn_out_aligned, "--retree 2", params$exe_mafft)
}

stopCluster(nwcl)
```

## Generate BUSCO trees
```{r check-busco-tree}
f_write_log(fn_log=fn_log, msg=c("", "------------ BUSCO Trees -----------"))
              
# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over shared BUSCOs
foreach (busco = shared_busco) %dopar% {
    dir_output <- paste0(dir_busco_tree, busco, "/")

    # output IQ-Tree2 files
    fn_fasta <- paste0(dir_output, busco, "_aligned.fa")
    fn_treefile <- paste0(dir_output, busco, "_aligned.fa.treefile")
    if (file.exists(fn_treefile) && !params$redo) {
        return(NULL)
    }

    # run IQ-Tree2
    f_iqtree2(fn_fasta, params$exe_iqtree2)
}

stopCluster(nwcl)
```

## Check BUSCO trees
```{r check-busco-tree-summary}
# select short reads
df_shortreads <- data.table::fread(params$file_shortreads, select="is_check")
n_shortreads <- df_shortreads$is_check[df_shortreads$is_check==TRUE]

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

ls_output <- foreach (busco = shared_busco, .combine='c') %dopar% {
    # initiate variable
    out <- list(busco=busco)

    # check if treefile exists
    fn_tree <- paste0(dir_busco_tree, busco, "/", busco, "_aligned.fa.treefile")
    if (!file.exists(fn_tree)){
        log4r::info(fn_logger, paste0("File not found: BUSCO tree for ", busco, ". Skipped."))
        return(NULL)
    }

    # read tree
    tre <- ape::read.tree(fn_tree)
    ls_tips <- tre$tip.label

    # iterate over reference sequences
    for (ref in ls_refseq) {
        # extract list of taxa with specific reference
        ls_taxa <- ls_tips[stringr::str_detect(ls_tips, ref)]
        if (length(ls_taxa) <= 1) {
            out[[ref]] <- "INVALID"
            next
        } else if (length(n_shortreads)+1 > length(ls_taxa)) {
            out[[ref]] <- "INCOMPLETE"
            next
        }

        # check if reference-based clade is monophyletic
        is_monophyletic <- ape::is.monophyletic(tre, tips = ls_taxa)
        if (is_monophyletic) {
            out[[ref]] <- "REFBIAS"
            next
        }

        out[[ref]] <- is_monophyletic
    }

    return(list(out))
}

stopCluster(nwcl)

# convert list into dataframe
df_output <- data.table::as.data.table(do.call(rbind, ls_output), fill=TRUE)
data.table::fwrite(df_output, file=fn_summary, sep="\t", quote=F, row.names=F)

log4r::info(fn_logger, "File created: summary for BUSCO trees.")
```