---
title: "Run Coordinate-based BUSCO Analysis on Reference Sequences and Short Reads"
---

```{r check-busco-setup, include=FALSE}
f_write_log(fn_log=fn_log,
            msg=c("", "####################################",
                      "####            BUSCO           ####",
                      "####################################", "",
                      "--------- COORDINATE-BASED ---------"))

# output directories
dir_busco <- paste0(dir_current, "/busco_check/busco_coordinates/")
dir_busco_lineage <- paste0(dir_current, "/busco_lineage/")

dir_busco_ref <- paste0(dir_busco, "refseq/")
dir_busco_read <- paste0(dir_busco, "short_reads/")
dir_busco_tree <- paste0(dir_busco, "trees/")

lapply(list(dir_busco_lineage,dir_busco_ref,dir_busco_read,dir_busco_tree), function(x){if(!dir.exists(x)) dir.create(x, recursive=T)})
```

```{r check-busco-params, include=FALSE}
thread_busco <- ifelse(params$thread < 50, params$thread, 50)
min_busco_depth <- 20
```

```{r busco-check-species-map, include=FALSE}
# create a map for ID and species name
ls_species_name <- c()

# read input file
df_refs <- data.table::fread(params$file_refseq)

# iterate over reference sequences
for (ref in 1:nrow(df_refs)) {
    ls_species_name[df_refs$id[ref]] <- df_refs$species[ref]
}

# iterate over short reads
for (read in 1:nrow(df_reads)) {
    ls_species_name[df_reads$id[read]] <- df_reads$species[read]
}
```

## Download BUSCO lineage dataset
```{r busco-lineage-download}
# download specific BUSCO lineage
dir_lineage <- paste0(dir_busco_lineage, "lineages/", params$busco_lineage)
if (!dir.exists(dir_lineage)) {
    busco_cmd <- paste(params$exe_busco, "--download_path", dir_busco_lineage, "--download", params$busco_lineage)
    system(busco_cmd)
}
```

## Run BUSCO on reference sequences
```{r, include=FALSE}
f_write_log(fn_log=fn_log, msg=c("", "--------- Reference Sequence -------"))

# set up variables
is_busco_ref <- FALSE

ls_busco_ref_done <- c()
ls_busco_ref_run <- c()

# iterate over reference sequences
if (params$redo) {
    ls_busco_ref_run <- ls_refseq
} else {
    for (ref in ls_refseq) {
        # check if BUSCO run exists
        file_busco_log <- paste0(dir_busco_ref, ref, "/logs/busco.log")
        if (file.exists(file_busco_log)) {
            finish_msg <- system(paste("grep 'BUSCO analysis done'", file_busco_log), intern=T)
            if (length(finish_msg) != 0) {
                ls_busco_ref_done <- c(ls_busco_ref_done, paste("-", ref))
                next
            }
        }

        # add reference sequence to to-do-list
        ls_busco_ref_run <- c(ls_busco_ref_run, ref)
    }
}

# output log file for available reference BUSCO runs
if (length(ls_busco_ref_done) > 0) {
    ls_busco_ref_done <- c(paste0("Available reference BUSCO runs (", length(ls_busco_ref_done), "/", length(ls_refseq), ")"), ls_busco_ref_done, "")
    f_write_log(fn_log=fn_log, msg=ls_busco_ref_done)
}

# update variable for the next analysis
if (length(ls_busco_ref_run) > 0) {
    is_busco_ref <- TRUE
}
```

```{r check-busco-ref, include=is_busco_ref, eval=is_busco_ref}
f_write_log(fn_log=fn_log, msg=paste("Running BUSCO using", thread_busco, "threads"))

# create doSNOW cluster
nwcl <- makeCluster(floor(params$thread/thread_busco))
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
foreach (ref = ls_refseq) %dopar% {
    # extract reference file location
    file_ref <- df_refseq_dir$dir[df_refseq_dir$id==ref]

    # run BUSCO
    f_run_busco(file_ref, dir_lineage, ref, dir_busco_ref, params$busco_mode, thread_busco, params$exe_busco)
    log4r::info(fn_logger, paste0("File created: BUSCO run for ", ref, "."))
}

stopCluster(nwcl)
```

## Extract common BUSCOs between reference sequences
```{r check-busco-common}
# initiate variable
shared_busco <- c()

# iterate over reference sequences
for (ref in ls_refseq) {
  # directory for BUSCO alignments
  dir_buscos <- paste0(dir_busco_ref, ref, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/")
  
  # remove file extension .fna
  busco_ids <- list.files(dir_buscos, pattern = "*.fna$", full.names = F, recursive = F)
  busco_ids <- sapply(busco_ids, function(x) { gsub(".fna", "", x) })
  
  # extract shared BUSCOs between reference sequences
  if (length(shared_busco) == 0) {
    shared_busco <- busco_ids
  } else {
    shared_busco <- intersect(shared_busco, busco_ids)
  }
}

# check if shared BUSCOs found
if (length(shared_busco) == 0) {
    log4r::error(fn_logger, paste0("Error: no BUSCO shared between reference sequences. Exited."))
    knitr::knit_exit()
} else {
    f_write_log(fn_log=fn_log, msg=c("", paste0("Number of complete single-copy BUSCOs: ", length(shared_busco), ".")))
}
```

## Extract BUSCOs for reference sequences
```{r check-busco-extract-ref}
# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
for (ref in ls_refseq) {
    # directory for BUSCO alignments
    dir_busco_out <- paste0(dir_busco_ref, ref, "/fasta/")
    if (!dir.exists(dir_busco_out)) {
        dir.create(dir_busco_out, recursive=T)
    }
  
    # extract reference file location
    file_fas <- df_refseq_dir$dir[df_refseq_dir$id==ref]

    # iterate over shared BUSCOs
    foreach (busco = shared_busco) %dopar% {
        require(data.table)
        require(Biostrings)
        require(stringr)

        # BUSCO files
        dir_busco_gff <- paste0(dir_busco_ref, ref, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/", busco)
        file_busco_fas <- paste0(dir_busco_gff, ".fna")
        file_busco_gff <- paste0(dir_busco_gff, ".gff")
        
        # extract header
        busco_header <- system(paste("grep '^>'", file_busco_fas), intern=T)

        # output files
        file_busco_cds <- paste0(dir_busco_out, busco, ".cds.fna")
        file_busco_gff_out <- paste0(dir_busco_out, busco, ".gff")
        file_busco_concat <- paste0(dir_busco_out, busco, ".fna")

        # check if file exists
        if (file.exists(file_busco_concat) && file.size(file_busco_concat) != 0L && !params$redo) {
            return(NULL)
        }

        # extract BUSCO coordinates
        coordinates <- f_extract_coordinates(busco_header, busco, ref)
        if (!is.null(coordinates$errmsg)) {
            f_write_log(fn_log=fn_log, msg=coordinates$errmsg)
            return(NULL)
        }

        # manipulate GFF file
        if (!file.exists(file_busco_gff)) {
            f_create_gff(coordinates, busco, file_busco_gff_out)
            log4r::warn(fn_logger, paste0("File not found: ", busco, " GFF file for ", ref, ". Extracting the whole alignment."))
        } else {
            msg <- f_manipulate_gff(file_busco_gff, coordinates, busco, ref, file_busco_gff_out)
            if (!is.null(msg)) {
                if (!is.null(msg$errmsg)) {
                    f_write_log(fn_log=fn_log, msg=msg$errmsg)
                    return(NULL)
                } else if (!is.null(msg$warnmsg)) {
                    f_write_log(fn_log=fn_log, msg=msg$warnmsg)
                }
            }
        }
        
        # extract BUSCO
        f_extract_fasta_from_gff(file_fas, file_busco_gff_out, file_busco_cds, file_busco_concat, params$exe_gffread)

        # sanity check if FASTA is the same
        output <- f_compare_fasta(file_busco_fas, file_busco_concat)
        if (!is.null(output$errmsg)) {
            log4r::warn(fn_logger, paste0("Warn: ", busco, " FASTA files for ", ref, ". Headers are non-identical."))
        } else {
            if (!output$is_identical){
                log4r::warn(fn_logger, paste0("Warn: ", busco, " FASTA file for ", ref, ". Sequences are non-identical (identity=", output$score, ")."))
            }
        }
    }

    log4r::info(fn_logger, paste0("File created/modified: BUSCO FASTA alignments for ", ref, "."))
}

stopCluster(nwcl)
```

## Extract BUSCOs for short reads
```{r check-busco-extract-reads}
f_write_log(fn_log=fn_log, msg=c("", "------------ Short Reads -----------"))

# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over short reads
for (prefix in ls_prefix) {
    # create outdir
    dir_busco_out <- paste0(dir_busco_read, prefix, "/")
    if (!dir.exists(dir_busco_out)) {
        dir.create(dir_busco_out, recursive=T)
    }

    # extract short read and reference sequence IDs
    ref <- unlist(strsplit(prefix, split="--"))[2]

    # check input files
    file_fas <- paste0(dir_readmap, prefix, "/", prefix, ".fa")
    if (!file.exists(file_fas)) {
        log4r::warn(fn_logger, paste0("File not found: FASTA file for ", prefix, ". Skipped."))
        next
    }

    # iterate over shared BUSCOs
    foreach (busco = shared_busco) %dopar% {
        require(Biostrings)
        require(stringr)

        # input files
        file_bam <- paste0(dir_readmap, prefix, "/", prefix, ".bam")
        file_busco_gff_out <- paste0(dir_busco_ref, ref, "/fasta/", busco, ".gff")
        if (!file.exists(file_busco_gff_out)) {
            log4r::warn(fn_logger, msg=paste0("File not found: ", busco, " GFF for ", ref, ". Skipped."))
            return(NULL)
        }

        # output files
        file_busco_cds <- paste0(dir_busco_out, busco, ".cds.fna")
        file_busco_concat <- paste0(dir_busco_out, busco, ".fna")

        # check if file exists
        if (file.exists(file_busco_concat) && file.size(file_busco_concat) != 0L && !params$redo) {
            return(NULL)
        }

        # calculate read depth
        read_coverage <- f_calculate_read_coverage(file_bam, file_busco_gff_out, params$exe_samtools)
        if (is.null(read_coverage)) {
            log4r::warn(fn_logger, msg=paste("BUSCO excluded:", busco, "for", prefix, "with no CDS"))
            return(NULL)
        } else if (read_coverage < min_busco_depth) {
            log4r::warn(fn_logger, msg=paste("BUSCO excluded:", busco, "for", prefix, "with read depth of", read_coverage))
            return(NULL)
        }

        # extract BUSCO
        f_extract_fasta_from_gff(file_fas, file_busco_gff_out, file_busco_cds, file_busco_concat, params$exe_gffread)
    }

    log4r::info(fn_logger, paste0("File created/modified: BUSCO FASTA alignments for ", prefix, "."))
}

stopCluster(nwcl)
```

## Convert individual BUSCO alignments into MSA
```{r check-busco-msa}
# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over shared BUSCOs
foreach (busco = shared_busco) %dopar% {
    dir_output <- paste0(dir_busco_tree, busco, "/")
    if (!dir.exists(dir_output)) {
        dir.create(dir_output, recursive=T)
    }

    # output MSA file
    fn_out <- paste0(dir_output, busco, ".fa")
    fn_out_aligned <- paste0(dir_output, busco, "_aligned.fa")
    if (all(file.exists(fn_out, fn_out_aligned)) && !params$redo) {
        # check if file is not empty
        if (file.size(fn_out_aligned) != 0L) {
            return(NULL)
        }
    }

    # extract BUSCO alignment from reference sequences
    for (ref in ls_refseq) {
        file_busco_ref <- paste0(dir_busco_ref, ref, "/fasta/", busco, ".fna")
        if (!file.exists(file_busco_ref)) {
            next
        }

        f_fasta2msa(file_busco_ref, ref, fn_out)
    }

    # extract BUSCO alignment from short reads
    for (read in ls_prefix) {
        file_busco_read <- paste0(dir_busco_read, read, "/", busco, ".fna")
        if (!file.exists(file_busco_read)) {
            next
        }

        f_fasta2msa(file_busco_read, read, fn_out)
    }

    # align using MAFFT FFT-NS-2
    f_mafft(fn_out, fn_out_aligned, "--retree 2", params$exe_mafft)
}

stopCluster(nwcl)
```

## Generate BUSCO trees
```{r check-busco-tree}
f_write_log(fn_log=fn_log, msg=c("", "------------ BUSCO Trees -----------"))
              
# create doSNOW cluster
nwcl <- makeCluster(params$thread)
doSNOW::registerDoSNOW(nwcl)

# iterate over shared BUSCOs
foreach (busco = shared_busco) %dopar% {
    dir_output <- paste0(dir_busco_tree, busco, "/")

    # output IQ-Tree2 files
    fn_fasta <- paste0(dir_output, busco, "_aligned.fa")
    fn_treefile <- paste0(dir_output, busco, "_aligned.fa.treefile")
    if (!file.exists(fn_treefile) || params$redo) {
        # run IQ-Tree2
        f_iqtree2(fn_fasta, params$exe_iqtree2)
    }

    # change the ID into species name
    fn_treefile_species <- paste0(dir_output, busco, "_aligned.fa.treefile.species")
    if (!file.exists(fn_treefile_species) || params$redo) {
        tre <- readLines(fn_treefile)
        for (id in names(ls_species_name)) {
            tre <- gsub(id, ls_species_name[id], tre)
        }
        writeLines(tre, con=fn_treefile_species)
    }
}

log4r::info(fn_logger, "File created/modified: BUSCO trees.")
stopCluster(nwcl)
```