---
title: "Calculate nRF on BUSCO Trees"
---

## Check correlation between number of references and topology distance
```{r check-busco-tree-setup, include=FALSE}
f_write_log(fn_log=fn_log,
            msg=c("", "####################################",
                      "####      Check BUSCO Trees     ####",
                      "####################################"))

# output directory
dir_check_tree <- paste0(dir_busco, "busco_tree/")
dir_check_tree_summary <- paste0(dir_check_tree, "summary/")
if (!dir.exists(dir_check_tree_summary)) {
    dir.create(dir_check_tree_summary, recursive=T)
}


# check input genome treefile
if (!file.exists(params$file_genome_treefile)) {
    log4r::error(fn_logger, "Genome treefile is not found. Exited.")
    knitr::knit_exit()
}

# output file
fn_busco_sumtable <- paste0(dir_check_tree_summary, "summary.dist.sumtable")
fn_busco_nref_tiff <- paste0(dir_check_tree_summary, "nRF_nref.tiff")
fn_busco_genomedist_tiff <- paste0(dir_check_tree_summary, "nRF_genomedist.tiff")
fn_busco_meandist_nref_tiff <- paste0(dir_check_tree_summary, "meandist_nref.tiff")
```

```{r, include=FALSE}
is_collapse_lowbs <- FALSE
```

```{r check-busco-tree-metadata}
# read genome treefile
genome_tree_id <- f_ref_tips_to_id(params$file_genome_treefile, df_refs)
genome_tree <- ape::read.tree(text=genome_tree_id)
genome_tree$edge.length <- ifelse(is.na(genome_tree$edge.length), 0.05, genome_tree$edge.length)
genome_dist <- ape::cophenetic.phylo(genome_tree)

# set number of threads
nthread <- ifelse(params$thread > length(shared_busco), length(shared_busco), params$thread)

# create doSNOW cluster
nwcl <- makeCluster(nthread)
doSNOW::registerDoSNOW(nwcl)

# iterate over number of reference sequences
for (i in 1:length(ls_refseq)) {
    # output directory
    dir_check_tree_category <- paste0(dir_check_tree, i, "/")
    if (!dir.exists(dir_check_tree_category)) {
        dir.create(dir_check_tree_category, recursive=T)
    }

    # output file
    fn_metadata <- paste0(dir_check_tree_category, "metadata.tsv")
    if (file.exists(fn_metadata) && !params$redo) {
        next
    }

    # iterate over BUSCOs
    ls_output <- foreach(busco = shared_busco, .combine='rbind') %dopar% {
        # create data.table to store metadata
        df_temp <- data.table::data.table(busco=character(), read=character(), ref=character(), genome_dist=numeric())

        # extract mapped reads
        df_mapped_reads_subset <- df_mapped_reads_busco[df_mapped_reads_busco$busco==busco & df_mapped_reads_busco$is_error==F,]
        ls_refseq_busco <- unique(df_mapped_reads_subset$ref)
        ls_shortreads_busco <- unique(df_mapped_reads_subset$read)

        # return NULL if number of references is inadequate
        if (length(ls_refseq_busco) < i) {
            return(NULL)
        }

        # extract random refs
        ls_sampled_refs <- sample(ls_refseq_busco, i, replace=FALSE)

        # iterate through the number of runs
        for (read in ls_shortreads_busco) {
            # extract reference ID based on short-reads
            read_species_name <- df_reads_ref$species[df_reads_ref$reads==read]
            read_ref_id <- df_reads_ref$ref[df_reads_ref$species==read_species_name]

            # extract the closest reference
            closest_ref <- f_check_closest_ref(genome_dist, read_ref_id, ls_sampled_refs)

            # update data.table
            df_temp <- rbind(df_temp, data.table::data.table(busco=busco, read=read, ref=closest_ref$ref, genome_dist=closest_ref$dist))
        }

        return(df_temp)
    }

    # save data.table
    data.table::fwrite(ls_output, file=fn_metadata, sep="\t", quote=F, row.names=F)
}

stopCluster(nwcl)
```

```{r check-busco-tree-alignment}
f_write_log(fn_log=fn_log, msg=c("", "---------- BUSCO Alignment ---------"))

# create doSNOW cluster
nwcl <- makeCluster(nthread)
doSNOW::registerDoSNOW(nwcl)

# iterate over number of reference sequences
for (i in 1:length(ls_refseq)) {
    # output directory
    dir_check_tree_category <- paste0(dir_check_tree, i, "/")

    # check if input file exists
    fn_metadata <- paste0(dir_check_tree_category, "metadata.tsv")
    if (!file.exists(fn_metadata)) {
        log4r::warn(fn_logger, paste("File not found: metadata for", i, "references. Skipped."))
        next
    }

    # extract unique BUSCOs
    df_metadata <- data.table::fread(fn_metadata)
    ls_uq_busco <- unique(df_metadata$busco)

    # iterate over BUSCOs
    foreach (busco = ls_uq_busco) %dopar% {
        # output directories
        dir_check_busco_tree <- paste0(dir_check_tree_category, busco, "/")
        dir_check_busco_tree_read <- paste0(dir_check_busco_tree, "read/")
        dir_check_busco_tree_refs <- paste0(dir_check_busco_tree, "refs/")
        lapply(list(dir_check_busco_tree_read,dir_check_busco_tree_refs), function(x){if(!dir.exists(x)) dir.create(x, recursive=T)})

        # output files
        fn_out_read <- paste0(dir_check_busco_tree_read, busco, ".fna")
        fn_out_refs <- paste0(dir_check_busco_tree_refs, busco, ".fna")
        fn_out_read_aligned <- paste0(dir_check_busco_tree_read, busco, "_aligned.fna")
        fn_out_refs_aligned <- paste0(dir_check_busco_tree_refs, busco, "_aligned.fna")

        # subset the metadata data.table
        df_metadata_subset <- df_metadata[df_metadata$busco == busco,]

        # check if output files exist for mapped reads
        if (!all(file.exists(fn_out_read, fn_out_read_aligned)) || params$redo) {
            # iterate over rows
            for (j in 1:nrow(df_metadata_subset)) {
                prefix <- paste0(df_metadata_subset$read[j], "--", df_metadata_subset$ref[j])

                # convert mapped reads ID to reference ID
                read_species_name <- df_reads_ref$species[df_reads_ref$reads==df_metadata_subset$read[j]]
                read_ref_id <- df_reads_ref$ref[df_reads_ref$species==read_species_name]

                # input FASTA file
                fn_fasta_read <- ""
                if (params$type == "coordinate" || params$type == "" || is.null(params$type)) {
                    fn_fasta_read <- paste0(dir_busco_read, prefix, "/", busco, ".fna")
                } else if (params$type == "pipeline") {
                    fn_fasta_read <- paste0(dir_busco_read, prefix, "/run_", params$busco_lineage, "/busco_sequences/single_copy_busco_sequences/", busco, ".fna")
                }

                # MSA
                f_fasta2msa(fn_fasta_read, read_ref_id, fn_out_read)
            }

            # align using MAFFT FFT-NS-2
            f_mafft(fn_out_read, fn_out_read_aligned, "--retree 2", params$exe_mafft)
        }

        # check if output files exist for reference sequences
        if (!all(file.exists(fn_out_refs, fn_out_refs_aligned)) || params$redo) {
            # iterate over rows
            for (j in 1:nrow(df_metadata_subset)) {
                # extract reference ID based on short-reads
                read_species_name <- df_reads_ref$species[df_reads_ref$reads==df_metadata_subset$read[j]]
                read_ref_id <- df_reads_ref$ref[df_reads_ref$species==read_species_name]

                # input FASTA file
                fn_fasta_refs <- paste0(dir_busco_ref, read_ref_id, "/fasta/", busco, ".fna")

                # MSA
                f_fasta2msa(fn_fasta_refs, read_ref_id, fn_out_refs)
            }

            # align using MAFFT FFT-NS-2
            f_mafft(fn_out_refs, fn_out_refs_aligned, "--retree 2", params$exe_mafft)
        }
    }
}

stopCluster(nwcl)
```

```{r check-busco-tree-build}
f_write_log(fn_log=fn_log, msg=c("", "------------- BUSCO Tree -----------"))

# create doSNOW cluster
nwcl <- makeCluster(nthread)
doSNOW::registerDoSNOW(nwcl)

# iterate over number of reference sequences
for (i in 1:length(ls_refseq)) {
    # output directory
    dir_check_tree_category <- paste0(dir_check_tree, i, "/")

    # list BUSCOs
    ls_uq_busco <- list.dirs(dir_check_tree_category, recursive=F, full.names=F)

    # iterate over BUSCOs
    foreach (busco = ls_uq_busco) %dopar% {
        # iterate over mapped reads and reference sequences
        lapply(list("read","refs"), function(type) {
            dir_output <- paste0(dir_check_tree_category, busco, "/", type, "/")

            # output IQ-Tree2 files
            fn_fasta <- paste0(dir_output, busco, "_aligned.fna")
            fn_treefile <- paste0(dir_output, busco, "_aligned.fna.treefile")
            if (!file.exists(fn_treefile) || params$redo) {
                # run IQ-Tree2
                f_iqtree2(fn_fasta, params$exe_iqtree2)
            }

            # output collapsed low-bootstrap branch
            fn_treefile_collapsed <- paste0(dir_output, busco, "_aligned.fna.collapsed.treefile")
            if (is_collapse_lowbs && (!file.exists(fn_treefile_collapsed) || params$redo)) {
                # run newick_utils
                f_collapse_branch(fn_treefile, 95, fn_treefile_collapsed, params$exe_nwed)
            }
        })
    }
}

stopCluster(nwcl)
```

```{r check-busco-tree-distance}
f_write_log(fn_log=fn_log, msg=c("", "-------- BUSCO Tree Distance -------"))

# output data.table
df_output <- data.table::data.table(n_ref=numeric(), busco=character(), ntaxa=numeric(), n_spmatch=numeric(), sum_genome_dist=numeric(), 
                                    nrf=numeric(), treedist=numeric(), treedist_highbs=numeric(), read_infsites=numeric(),
                                    ratio_treelen=numeric(), ratio_meanbs=numeric(), ratio_read_tiplenlen=numeric())

# create doSNOW cluster
nwcl <- makeCluster(nthread)
doSNOW::registerDoSNOW(nwcl)

# iterate over number of reference sequences
for (i in 1:length(ls_refseq)) {
    # output directory
    dir_check_tree_category <- paste0(dir_check_tree, i, "/")

    # open metadata file
    fn_metadata <- paste0(dir_check_tree_category, "metadata.tsv")
    if (!file.exists(fn_metadata)) {
        next
    }

    # input files
    df_metadata <- data.table::fread(fn_metadata)
    ls_uq_busco <- list.dirs(dir_check_tree_category, recursive=F, full.names=F)

    # iterate over BUSCOs
    df_output_subset <- foreach (busco = ls_uq_busco, .combine='rbind') %dopar% {
        # input files
        dir_check_busco_tree <- paste0(dir_check_tree_category, busco, "/")
        fn_tree_one <- paste0(dir_check_busco_tree, "read/", busco, "_aligned.fna.treefile")
        fn_tree_two <- paste0(dir_check_busco_tree, "refs/", busco, "_aligned.fna.treefile")
        
        # check if file exists
        if (!all(file.exists(fn_tree_one, fn_tree_two))) {
            return(NULL)
        }

        # check the number of taxa
        tre <- ape::read.tree(fn_tree_two)

        # extract the total genome distance
        genome_dist <- sum(df_metadata$genome_dist[df_metadata$busco==busco])
        n_spmatch <- length(df_metadata$read[df_metadata$busco==busco & df_metadata$genome_dist==0])

        # extract the number of parsimony informative sites
        fn_log_one <- paste0(dir_check_busco_tree, "read/", busco, "_aligned.fna.iqtree")
        fn_log_two <- paste0(dir_check_busco_tree, "refs/", busco, "_aligned.fna.iqtree")

        # extract tree lengths
        read_blen <- as.numeric(gsub("^.* ", "", system(paste("grep '^Total tree length (sum of branch lengths):'",fn_log_one), intern = T)))
        refs_blen <- as.numeric(gsub("^.* ", "", system(paste("grep '^Total tree length (sum of branch lengths):'",fn_log_two), intern = T)))
        ratio_treelen <- round(read_blen/refs_blen, 3)

        # extract ratio of tip to total tree lengths
        read_inblen <- gsub("^.*\\: ", "", system(paste("grep '^Sum of internal branch lengths:'",fn_log_one), intern = T))
        read_inblen <- as.numeric(gsub(".\\(.*$", "", read_inblen))
        ratio_read_tiplenlen <- 1 - round(read_inblen/read_blen, 3)

        # extract number of informative sites
        n_infsites <- gsub("^.* ", "", system(paste("grep '^Number of parsimony informative sites:'",fn_log_one), intern = T))

        # extract average bootstrap
        read_meanbs <- f_calculate_mean_bs(fn_tree_one)
        refs_meanbs <- f_calculate_mean_bs(fn_tree_two)
        ratio_meanbs <- round(read_meanbs/refs_meanbs, 3)

        # calculate nRF
        nrf_dist <- f_calculate_nRF(fn_tree_one, fn_tree_two)
        treedist <- f_calculate_treedist(fn_tree_one, fn_tree_two, 0)
        treedist_highbs <- f_calculate_treedist(fn_tree_one, fn_tree_two, 95)

        return(data.table::data.table(n_ref=i, busco=busco, ntaxa=length(tre$tip.label), n_spmatch=n_spmatch, sum_genome_dist=genome_dist, 
                                      nrf=nrf_dist, treedist=treedist$dist, treedist_highbs=treedist_highbs$dist, read_infsites=n_infsites,
                                      ratio_treelen=ratio_treelen, ratio_meanbs=ratio_meanbs, ratio_read_tiplenlen=ratio_read_tiplenlen))
    }

    # update data.table
    df_output <- rbind(df_output, df_output_subset)
}

stopCluster(nwcl)

# save file
data.table::fwrite(df_output, file=fn_busco_sumtable, quote=F, sep="\t")

# nRF vs. total genomic distance
plot <- ggplot(df_output, aes(x=sum_genome_dist, y=nrf, ymin=0, ymax=1)) +
    geom_point(size=2) +
    ggtitle("Normalised RF Distance across Total Genomic Distances") + xlab("Genomic distance") + ylab("nRF distance") +
    facet_wrap(.~n_ref) +
    theme(plot.title = element_text(hjust = 0.5, size = 50),
          plot.margin = margin(1.25, 1.25, 1.25, 1.25, "cm"),
          axis.title.x=element_text(size = 40, margin = margin(t=20, r=0, b=0, l=0)),
          axis.title.y=element_text(size = 40, margin = margin(t=0, r=20, b=0, l=0)),
          axis.text.y=element_text(size=30),
          axis.text.x=element_text(size=30),
          strip.text = element_text(size=30),
          legend.title=element_text(size=30),
          legend.text=element_text(size=30),
          legend.key.size=unit(2,"cm"))

tiff(file=fn_busco_genomedist_tiff, units="px", width=2880, height=1800)
print(plot)
dev.off()

# nRF vs. number of reference sequences
plot <- ggplot(df_output, aes(x=as.factor(n_ref), y=nrf, ymin=0, ymax=1)) +
    geom_point(position=position_jitter(height=0.01, width=0.05), size=2, show.legend=F) +
    geom_violin(alpha=0.5) +
    ggtitle("Normalised RF Distance across Number of References") + xlab("Number of references") + ylab("nRF distance") +
    theme(plot.title = element_text(hjust = 0.5, size = 50),
          plot.margin = margin(1.25, 1.25, 1.25, 1.25, "cm"),
          axis.title.x=element_text(size = 40, margin = margin(t=20, r=0, b=0, l=0)),
          axis.title.y=element_text(size = 40, margin = margin(t=0, r=20, b=0, l=0)),
          axis.text.y=element_text(size=30),
          axis.text.x=element_text(size=30),
          strip.text = element_text(size=30),
          legend.title=element_text(size=30),
          legend.text=element_text(size=30),
          legend.key.size=unit(2,"cm"))

tiff(file=fn_busco_nref_tiff, units="px", width=2880, height=1800)
print(plot) 
dev.off()

# mean distances vs. number of reference sequences
df_output_mean <- df_output %>%
                    group_by(n_ref) %>%
                    summarise(nrf=mean(nrf), treedist=mean(treedist), treedist_highbs=mean(treedist_highbs)) %>%
                    tidyr::pivot_longer(-n_ref)

plot <- ggplot(df_output_mean, aes(x=as.factor(n_ref), y=value, colour=name, group=name, ymin=0, ymax=1)) +
    geom_line(linewidth=2) +
    geom_point(size=2) +
    ggtitle("Mean of Tree Distances across Number of References") + xlab("Number of references") + ylab("Mean distance") +
    theme(plot.title=element_text(hjust = 0.5, size = 50),
          plot.margin=margin(1.25, 1.25, 1.25, 1.25, "cm"),
          axis.title.x=element_text(size = 40, margin = margin(t=20, r=0, b=0, l=0)),
          axis.title.y=element_text(size = 40, margin = margin(t=0, r=20, b=0, l=0)),
          axis.text.y=element_text(size=30),
          axis.text.x=element_text(size=30),
          strip.text=element_text(size=30),
          legend.title=element_blank(),
          legend.text=element_text(size=30),
          legend.key.size=unit(2,"cm"))

tiff(file=fn_busco_meandist_nref_tiff, units="px", width=2880, height=1800)
print(plot)
dev.off()
```

```{r, out.width="100%", eval=FALSE}
# note: add filters to differentiate problematic BUSCOs
print(plot + theme(plot.title = element_text(face = "bold"),
                   plot.margin = margin(0.5, 0.5, 0.5, 0.5, "cm")))
```