---
title: "Read Mapping of Short Reads"
---

```{r readmap-setup, include=FALSE}
write.table(c("",
              "####################################",
              "####        Read Mapping        ####",
              "####################################"),
            file=fn_log, quote=F, row.names=F, col.names=F, append=T)

# create readmap directory
dir_readmap <- paste0(getwd(), "/readmap/")
if (!dir.exists(dir_readmap)) {
  dir.create(dir_readmap, recursive = T)
}

# input/output files
file_mashsum <- paste0(getwd(), "/", params$prefix, ".mashsum")
```

```{r readmap}
# read mashsum file
df_mashsum <- data.table::fread(file_mashsum)

# iterate over all short reads
for (i in 1:nrow(df_mashsum)) {
    dir_shortread_id <- paste0(dir_readmap, df_mashsum$reads_accession[i], "/")
    if (!dir.exists(dir_shortread_id)) {
        dir.create(dir_shortread_id, recursive=T)
    }

    file_sam <- paste0(dir_shortread_id, df_mashsum$reads_accession[i], ".sam")
    if (file.exists(file_sam) && !params$redo) {
        log4r::warn(fn_logger, paste0("File found: SAM file for ", df_mashsum$reads_accession[i], ". Skipped."))
        next
    }
    
    # check the number of fastq files
    ls_dir_reads <- unlist(strsplit(df_mashsum$reads_dir[i], split="/"))
    dir_reads <- paste(ls_dir_reads[-length(ls_dir_reads)], collapse="/")

    ls_fastq <- list.files(dir_reads, pattern="*.fastq", recursive=F, full.names=T)
    len_ls_fastq <- length(ls_fastq)
    if (len_ls_fastq != 1 && len_ls_fastq != 2) {
        log4r::warn(fn_logger, paste0("Warn: invalid number of FASTQ file for ", df_mashsum$reads_accession[i], " (", len_ls_fastq, "). Skipped."))
        next
    } else {
        f_read_map(df_mashsum$refseq_dir, ls_fastq, params$thread, params$exe_bwamem2, file_sam)
        log4r::info(fn_logger, paste0("File created: SAM file for ", df_mashsum$reads_accession[i], "."))
    }
}
```

```{r file-conversion}
# iterate over all short reads
for (i in 1:nrow(df_mashsum)) {
    # outdir
    dir_shortread_id <- paste0(dir_readmap, df_mashsum$reads_accession[i], "/")

    # convert SAM to BAM and FASTA
    f_convert_SAM(df_mashsum$reads_accession[i], dir_shortread_id, params$thread, params$exe_samtools)

    log4r::info(fn_logger, paste0("File created: BAM and FASTA files for ", prefix, "."))
}
```