---
title: "Read Mapping of Short Reads on Reference Sequences"
---

```{r, include=FALSE}
f_write_log(fn_log=fn_log,
            msg=c("", "####################################",
                      "####        Read Mapping        ####",
                      "####################################"))

# list of prefix
ls_prefix <- as.vector(outer(ls_shortreads, ls_refseq, FUN = function(x, y) paste(x, y, sep = "--")))

# output directories
dir_shortreads_filtered <- paste0(dir_shortreads, "/filtered/")
dir_readmap <- paste0(currentdir, "/readmap/")

# create directory if not exist
lapply(list(dir_shortreads_filtered,dir_readmap), function(x){if(!dir.exists(x)) dir.create(x, recursive=T)})

# output files
fn_readmap_metadata <- paste0(dir_readmap, "metadata.tsv")
```

## QC of short reads
```{r, include=FALSE}
f_write_log(fn_log=fn_log, msg=c("", "----- Quality Control of FASTQ -----"))

# set up variables
is_qcreads <- FALSE

ls_shortread_done <- c()
ls_shortread_qc <- c()

# iterate over short reads
if (params$redo) {
    ls_shortread_qc <- ls_shortreads
} else {
    for (read in ls_shortreads) {
        # check if output file exists
        file_fastq <- paste0(dir_shortreads_filtered, read, ".pair1.truncated")
        if (file.exists(file_fastq)) {
            ls_shortread_done <- c(ls_shortread_done, paste("-", read))
            next
        }

        # add short reads to to-do-list
        ls_shortread_qc <- c(ls_shortread_qc, read)
    }
}

# output log file for available filtered short reads
if (length(ls_shortread_done) > 0) {
    ls_shortread_done <- c(paste0("Available filtered short reads (", length(ls_shortread_done), "/", length(ls_shortreads), ")"), ls_shortread_done, "")
    f_write_log(fn_log=fn_log, msg=ls_shortread_done)
}

# update variable for the next analysis
if (length(ls_shortread_qc) > 0) {
    is_qcreads <- TRUE
}
```

```{r}
# print the number of filtered short reads in HTML document
cat(paste0("Available filtered short reads (", length(ls_shortread_done), "/", length(ls_shortreads), ")"))
cat(paste("Short reads to be filtered:", length(ls_shortread_qc)))
```

```{r readmap-qcreads, include=is_qcreads, eval=is_qcreads}
f_write_log(fn_log=fn_log, msg=paste("Running AdapterRemoval using", params$thread_adapterremoval, "threads"))

# create doSNOW cluster
cl_adapterremoval <- floor(params$thread/params$thread_adapterremoval)
if (cl_adapterremoval < 1) {
    log4r::error(fn_logger, "Invalid thread for AdapterRemoval. Exited.")
    knitr::knit_exit()
}

nwcl <- makeCluster(cl_adapterremoval)
doSNOW::registerDoSNOW(nwcl)

# iterate over short reads
foreach (read = ls_shortread_qc) %dopar% {
    prefix <- paste0(dir_shortreads_filtered, read)

    # extract FASTQ files
    dir_reads <- paste0(dir_shortreads, read, "/fastq/")
    fn_fastq_one <- paste0(dir_reads, read, "_1.fastq")
    fn_fastq_two <- paste0(dir_reads, read, "_2.fastq")
    
    # check the forward FASTQ file
    if (!file.exists(fn_fastq_one)) {
        log4r::warn(fn_logger, paste0("File not found: FASTQ file for ", read, ". Skipped."))
        return(NULL)
    }

    # check the reverse FASTQ file
    if (!file.exists(fn_fastq_two)) {
        fn_fastq_two <- NULL
    }
    
    # run AdapterRemoval
    f_qc_short_reads(fn_fastq_one, fn_fastq_two, params$file_adapters, prefix, params$min_read_quality, params$thread_adapterremoval, params$exe_adapterremoval)
    log4r::info(fn_logger, paste0("File created: filtered FASTQ file for ", read, "."))
}

stopCluster(nwcl)
```

## Map short reads to reference using BWA-MEM2
```{r readmap-refmap, include=FALSE}
# create data.frame to store directory of reference sequence
df_refseq_dir <- data.frame(id=character(), dir=character())

# iterate over reference sequence
for (ref in ls_refseq) {
    # extract reference sequence
    dir_ref <- paste0(dir_refseq, ref, "/ncbi_dataset/data/", ref, "/")
    file_ref <- list.files(dir_ref, pattern="*.fna$", recursive=F, full.names=T)

    # check if there are multiple reference sequences
    if (length(file_ref) > 1) {
        file_ref <- file_ref[1]
        log4r::warn(fn_logger, paste0("Multiple references are found for ", ref, ". ", file_ref, " is selected."))
    }

    # update data.frame
    df_refseq_dir <- rbind(df_refseq_dir, data.frame(id=ref, dir=file_ref))
}

data.table::fwrite(df_refseq_dir, file=fn_readmap_metadata, sep="\t", quote=F)
```

```{r, include=FALSE}
f_write_log(fn_log=fn_log, msg=c("", "------------ Read Mapping ----------"))

# set up variables
is_readmap <- FALSE

ls_readmap_done <- c()
ls_readmap_run <- c()

# iterate over pairs of reference sequence and short reads
if (params$redo) {
    ls_readmap_run <- ls_prefix
} else {
    for (prefix in ls_prefix) {
        # check if file exists
        file_bam <- paste0(dir_readmap, prefix, "/", prefix, ".bam")
        if (file.exists(file_bam)) {
            ls_readmap_done <- c(ls_readmap_done, paste("-", prefix))
            next
        }

        # add prefix to to-do-list
        ls_readmap_run <- c(ls_readmap_run, prefix)
    }
}

# output log file for available mapped reads
if (length(ls_readmap_done) > 0) {
    ls_readmap_done <- c(paste0("Available mapped reads (", length(ls_readmap_done), "/", length(ls_prefix), ")"), ls_readmap_done, "")
    f_write_log(fn_log=fn_log, msg=ls_readmap_done)
}

# update variable for the next analysis
if (length(ls_readmap_run) > 0) {
    is_readmap <- TRUE
}
```

```{r}
# print the number of mapped reads in HTML document
cat(paste0("Available mapped reads (", length(ls_readmap_done), "/", length(ls_prefix), ")"))
cat(paste("Short reads to be mapped:", length(ls_readmap_run)))
```

```{r readmap, include=is_readmap, eval=is_readmap}
f_write_log(fn_log=fn_log, msg=paste("Running BWA-MEM2 using", params$thread_bwamem2, "threads"))

# create doSNOW cluster
cl_bwamem2 <- floor(params$thread/params$thread_bwamem2)
if (cl_bwamem2 < 1) {
    log4r::error(fn_logger, "Invalid thread for BWA-MEM2. Exited.")
    knitr::knit_exit()
}

# create doSNOW cluster
nwcl <- makeCluster(cl_bwamem2)
doSNOW::registerDoSNOW(nwcl)

# iterate over prefix
foreach (prefix = ls_readmap_run) %dopar% {
    # output directory
    dir_output <- paste0(dir_readmap, prefix, "/")
    if (!dir.exists(dir_output)) {
        dir.create(dir_output, recursive=T)
    }

    # output file
    file_sam <- paste0(dir_output, prefix, ".sam")
    file_bam <- paste0(dir_output, prefix, ".bam")

    # extract reference and short reads
    read <- unlist(strsplit(prefix, split="--"))[1]
    ref <- unlist(strsplit(prefix, split="--"))[2]

    # extract reference file location
    file_ref <- df_refseq_dir$dir[df_refseq_dir$id==ref]

    # extract FASTQ file
    file_fastq_one <- paste0(dir_shortreads_filtered, read, ".pair1.truncated")
    if (!file.exists(file_fastq_one)) {
        log4r::warn(fn_logger, paste0("File not found: filtered FASTQ file for ", read, ". Skipped."))
        return(NULL)
    }

    file_fastq_two <- paste0(dir_shortreads_filtered, read, ".pair2.truncated")
    if (!file.exists(file_fastq_two)) {
        file_fastq_two <- NULL
    }
    
    # run BWA-MEM2
    f_read_mapping(file_ref, file_fastq_one, file_fastq_two, params$thread_bwamem2, params$exe_bwamem2, file_sam)

    # convert SAM to BAM
    f_sam_to_bam(prefix, dir_output, params$thread_samtools, params$exe_samtools)

    # delete SAM file
    unlink(file_sam)
    log4r::info(fn_logger, paste0("File created: BAM file for ", prefix, "."))
}

stopCluster(nwcl)
```

```{r readmap-coverage, include=FALSE}
f_write_log(fn_log=fn_log, msg=paste("Running QualiMap using", params$thread_qualimap, "threads"))

# set up variable
ls_coverage_run <- c()

# iterate over prefix
if (params$redo) {
    ls_coverage_run <- ls_prefix
} else {
    for (prefix in ls_prefix) {
        # check if file exists
        file_qualimap_html <- paste0(dir_readmap, prefix, "/qualimap/qualimapReport.html")
        if (!file.exists(file_qualimap_html)) {
            ls_coverage_run <- c(ls_coverage_run, prefix)
        }
    }
}

# create doSNOW cluster
cl_qualimap <- floor(params$thread/params$thread_qualimap)
if (cl_qualimap < 1) {
    log4r::error(fn_logger, "Invalid thread for QualiMap. Exited.")
    knitr::knit_exit()
}

nwcl <- makeCluster(cl_qualimap)
doSNOW::registerDoSNOW(nwcl)

# iterate over mapped reads
df_coverage <- foreach (prefix = ls_prefix, .combine='rbind') %dopar% {
    # input parameters
    dir_qualimap_output <- paste0(dir_readmap, prefix, "/qualimap/")
    file_bam <- paste0(dir_readmap, prefix, "/", prefix, ".bam")

    # run QualiMap
    if (prefix %in% ls_coverage_run) {
        f_qualimap(file_bam, dir_qualimap_output, params$thread_qualimap, NULL, params$exe_qualimap)
    }

    # extract coverage
    mean_coverage <- system(paste0("grep 'mean coverageData' ", dir_qualimap_output, "genome_results.txt"), intern=T)
    mean_coverage <- strsplit(mean_coverage, split=" = ")[[1]][2]
    mean_coverage <- as.numeric(strsplit(mean_coverage, split="X")[[1]][1])

    # extract percentage of mapped reads
    mapped_reads <- system(paste0("grep 'number of mapped reads' ", dir_qualimap_output, "genome_results.txt"), intern=T)
    mapped_reads <- stringr::str_match(mapped_reads, "\\((.*)\\)")[,2]
    mapped_reads <- as.numeric(strsplit(mapped_reads, split="%")[[1]][1])

    # extract reference and reads
    readref <- strsplit(prefix, split="--")[[1]]

    return(data.table::data.table(read_id=readref[1],
                                  ref_id=readref[2],
                                  read_sp=df_reads$species[df_reads$id==readref[1]],
                                  ref_sp=df_refs$species[df_refs$id==readref[2]],
                                  mean_coverage=mean_coverage,
                                  mapped_reads=mapped_reads))
}

stopCluster(nwcl)

# save the file
fn_output <- paste0(dir_readmap, "summary.tsv")
data.table::fwrite(df_coverage, file=fn_output, sep="\t", quote=F, row.names=F, col.names=F)
```

## Variant calling
```{r, include=FALSE}
f_write_log(fn_log=fn_log, msg=c("", "---------- Variant Calling ---------"))

# set up variables
is_variant <- FALSE

ls_variant_done <- c()
ls_variant_run <- c()

# iterate over prefix
if (params$redo) {
    ls_variant_run <- ls_prefix
} else {
    for (prefix in ls_prefix) {
        # check if file exists
        file_vcf <- paste0(dir_readmap, prefix, "/", prefix, ".vcf.gz")
        file_fas <- paste0(dir_readmap, prefix, "/", prefix, ".fa")
        if (all(file.exists(file_vcf, file_fas)) && file.size(file_fas) != 0L) {
            ls_variant_done <- c(ls_variant_done, paste("-", prefix))
            next
        }

        # add prefix to to-do-list
        ls_variant_run <- c(ls_variant_run, prefix)
    }
}

# output log file for available VCF files
if (length(ls_variant_done) > 0) {
    ls_variant_done <- c(paste0("Available VCF files (", length(ls_variant_done), "/", length(ls_prefix), ")"), ls_variant_done, "")
    f_write_log(fn_log=fn_log, msg=ls_variant_done)
}

# update variable for the next analysis
if (length(ls_variant_run) > 0) {
    is_variant <- TRUE
}
```

```{r}
# print the number of VCF files in HTML document
cat(paste0("Available VCF files (", length(ls_variant_done), "/", length(ls_prefix), ")"))
cat(paste("Variants to be called:", length(ls_variant_run)))
```

```{r readmap-calling, include=is_variant, eval=is_variant}
f_write_log(fn_log=fn_log, msg=paste("Running Bcftools using", params$thread_samtools, "threads"))

# create doSNOW cluster
cl_samtools <- floor(params$thread/params$thread_samtools)
if (cl_samtools < 1) {
    log4r::error(fn_logger, "Invalid thread for Samtools. Exited.")
    knitr::knit_exit()
}

nwcl <- makeCluster(cl_samtools)
doSNOW::registerDoSNOW(nwcl)

# iterate over reference sequences
foreach (prefix = ls_variant_run) %dopar% {
    # output directory
    dir_output <- paste0(dir_readmap, prefix, "/")

    # check if input file exists
    file_bam <- paste0(dir_output, prefix, ".bam")
    if (!file.exists(file_bam)) {
        log4r::warn(fn_logger, paste0("File not found: BAM file for ", prefix, ". Skipped."))
        return(NULL)
    }

    # extract reference file location
    ref <- unlist(strsplit(prefix, split="--"))[2]
    file_ref <- df_refseq_dir$dir[df_refseq_dir$id==ref]

    # variant calling
    f_variant_calling(prefix, dir_output, params$thread_samtools, file_ref, params$exe_bcftools)
    log4r::info(fn_logger, paste0("File created: VCF and FASTA files for ", prefix, "."))
}

stopCluster(nwcl)
```